{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Language Model Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import openai\n",
    "import os\n",
    "\n",
    "openai.api_key =  os.getenv('OPENAI_API_KEY')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.chat_models import ChatOpenAI\n",
    "llm = ChatOpenAI(temperature=0.0, model_name='gpt-3.5-turbo')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tokenizer Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Encoding 'cl100k_base'>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import tiktoken \n",
    "tiktoken.encoding_for_model('gpt-3.5-turbo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tokenizer = tiktoken.get_encoding('cl100k_base')\n",
    "\n",
    "def tiktoken_len(text):\n",
    "    tokens = tokenizer.encode(\n",
    "        text,\n",
    "        disallowed_special=()\n",
    "    )\n",
    "    return len(tokens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.embeddings.openai import OpenAIEmbeddings\n",
    "\n",
    "model_name = 'text-embedding-ada-002'\n",
    "\n",
    "embed = OpenAIEmbeddings(\n",
    "    model=model_name,\n",
    "    openai_api_key=os.environ['OPENAI_API_KEY']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "text_examples = [\"I am a text example\", \n",
    "                 \"I am another text example\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = embed.embed_documents(text_examples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1536"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(result[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importing Raw Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_medicus_drug = 'raw_datasets/kegg_medicus_drug_en.csv'\n",
    "kegg_medicus_disease = 'raw_datasets/kegg_medicus_disease_en.csv'\n",
    "kegg_medicus_dgroup = 'raw_datasets/kegg_medicus_dgroup_en.csv'\n",
    "kegg_medicus_environ = 'raw_datasets/kegg_medicus_environ_en.csv'\n",
    "kegg_medicus_network = 'raw_datasets/kegg_medicus_network.csv'\n",
    "kegg_medicus_variant = 'raw_datasets/kegg_medicus_variant.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "raw_drug_df = pd.read_csv(kegg_medicus_drug)\n",
    "raw_disease_df = pd.read_csv(kegg_medicus_disease)\n",
    "raw_dgroup_df = pd.read_csv(kegg_medicus_dgroup)\n",
    "raw_environ_df = pd.read_csv(kegg_medicus_environ)\n",
    "raw_network_df = pd.read_csv(kegg_medicus_network)\n",
    "raw_variant_df = pd.read_csv(kegg_medicus_variant)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Cleaning & Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# empty variable for preprocessing function\n",
    "row_to_dict = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "preprocessed_datasets = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_dataframe_to_list(df):\n",
    "    result_list = df.apply(row_to_dict, axis=1).tolist()\n",
    "    return result_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kegg Medicus Drug Data Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry_id</th>\n",
       "      <th>classification</th>\n",
       "      <th>name</th>\n",
       "      <th>component</th>\n",
       "      <th>formula</th>\n",
       "      <th>exact_mass</th>\n",
       "      <th>mol_weight</th>\n",
       "      <th>image</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sequence type</th>\n",
       "      <th>...</th>\n",
       "      <th>dblinks_w_link</th>\n",
       "      <th>comment_w_link</th>\n",
       "      <th>interaction_w_link</th>\n",
       "      <th>target_w_link</th>\n",
       "      <th>source_w_link</th>\n",
       "      <th>metabolism_w_link</th>\n",
       "      <th>sequence_w_link</th>\n",
       "      <th>component_w_link</th>\n",
       "      <th>class_w_link</th>\n",
       "      <th>raw_entry_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D00001</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Water (JP18/USP);\\nPurified water (JP18);\\nPur...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H2O</td>\n",
       "      <td>18.0106</td>\n",
       "      <td>18.0153</td>\n",
       "      <td>D00001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CAS&lt;a href=\"https://identifiers.org/cas:7732-1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D00002</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Nadide (JAN/USAN/INN);\\nNicotinamide adenine d...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C21H28N7O14P2</td>\n",
       "      <td>664.1169</td>\n",
       "      <td>664.4330</td>\n",
       "      <td>D00002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CAS&lt;a href=\"https://identifiers.org/cas:53-84-...</td>\n",
       "      <td>Antagonist (to alcohol and narcotics)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D00003</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Oxygen (JP18/USP)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O2</td>\n",
       "      <td>31.9898</td>\n",
       "      <td>31.9988</td>\n",
       "      <td>D00003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CAS&lt;a href=\"https://identifiers.org/cas:7782-4...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D00004</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Carbon dioxide (JP18/USP);\\nCarbon dioxide (TN)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CO2</td>\n",
       "      <td>43.9898</td>\n",
       "      <td>44.0095</td>\n",
       "      <td>D00004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CAS&lt;a href=\"https://identifiers.org/cas:124-38...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D00004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D00005</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Flavin adenine dinucleotide (JAN);\\nAdeflavin ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C27H33N9O15P2</td>\n",
       "      <td>785.1571</td>\n",
       "      <td>785.5497</td>\n",
       "      <td>D00005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CAS&lt;a href=\"https://identifiers.org/cas:146-14...</td>\n",
       "      <td>Coenzyme form of vitamin B2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D00005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  entry_id classification                                               name  \\\n",
       "0   D00001           Drug  Water (JP18/USP);\\nPurified water (JP18);\\nPur...   \n",
       "1   D00002           Drug  Nadide (JAN/USAN/INN);\\nNicotinamide adenine d...   \n",
       "2   D00003           Drug                                  Oxygen (JP18/USP)   \n",
       "3   D00004           Drug    Carbon dioxide (JP18/USP);\\nCarbon dioxide (TN)   \n",
       "4   D00005           Drug  Flavin adenine dinucleotide (JAN);\\nAdeflavin ...   \n",
       "\n",
       "  component        formula  exact_mass  mol_weight   image sequence  \\\n",
       "0       NaN            H2O     18.0106     18.0153  D00001      NaN   \n",
       "1       NaN  C21H28N7O14P2    664.1169    664.4330  D00002      NaN   \n",
       "2       NaN             O2     31.9898     31.9988  D00003      NaN   \n",
       "3       NaN            CO2     43.9898     44.0095  D00004      NaN   \n",
       "4       NaN  C27H33N9O15P2    785.1571    785.5497  D00005      NaN   \n",
       "\n",
       "  sequence type  ...                                     dblinks_w_link  \\\n",
       "0           NaN  ...  CAS<a href=\"https://identifiers.org/cas:7732-1...   \n",
       "1           NaN  ...  CAS<a href=\"https://identifiers.org/cas:53-84-...   \n",
       "2           NaN  ...  CAS<a href=\"https://identifiers.org/cas:7782-4...   \n",
       "3           NaN  ...  CAS<a href=\"https://identifiers.org/cas:124-38...   \n",
       "4           NaN  ...  CAS<a href=\"https://identifiers.org/cas:146-14...   \n",
       "\n",
       "                          comment_w_link interaction_w_link target_w_link  \\\n",
       "0                                    NaN                NaN           NaN   \n",
       "1  Antagonist (to alcohol and narcotics)                NaN           NaN   \n",
       "2                                    NaN                NaN           NaN   \n",
       "3                                    NaN                NaN           NaN   \n",
       "4            Coenzyme form of vitamin B2                NaN           NaN   \n",
       "\n",
       "  source_w_link metabolism_w_link sequence_w_link component_w_link  \\\n",
       "0           NaN               NaN             NaN              NaN   \n",
       "1           NaN               NaN             NaN              NaN   \n",
       "2           NaN               NaN             NaN              NaN   \n",
       "3           NaN               NaN             NaN              NaN   \n",
       "4           NaN               NaN             NaN              NaN   \n",
       "\n",
       "  class_w_link raw_entry_id  \n",
       "0          NaN       D00001  \n",
       "1          NaN       D00002  \n",
       "2          NaN       D00003  \n",
       "3          NaN       D00004  \n",
       "4          NaN       D00005  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_drug_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "classification\n",
       "Crude Drug      1\n",
       "Drug            1\n",
       "Formula Drug    1\n",
       "Mixture Drug    1\n",
       "Name: classification, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_drug_df.groupby('classification')['classification'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "entry_id                  0\n",
       "classification            0\n",
       "name                      1\n",
       "component             10645\n",
       "formula                2296\n",
       "exact_mass             2658\n",
       "mol_weight             2658\n",
       "image                     0\n",
       "sequence              11227\n",
       "sequence type         11225\n",
       "source                11325\n",
       "class                  6430\n",
       "remark                 4109\n",
       "efficacy                 24\n",
       "disease               10516\n",
       "comment                5738\n",
       "target                 5507\n",
       "metabolism            10733\n",
       "interaction           11591\n",
       "db_links                244\n",
       "kcf                    3139\n",
       "atom                   3139\n",
       "bond                   3139\n",
       "bracket               10980\n",
       "original              10986\n",
       "repeat                10986\n",
       "disease_w_link        10522\n",
       "remark_w_link          4109\n",
       "efficacy_w_link       11938\n",
       "dblinks_w_link          244\n",
       "comment_w_link         5738\n",
       "interaction_w_link     6208\n",
       "target_w_link          5507\n",
       "source_w_link         11325\n",
       "metabolism_w_link     10733\n",
       "sequence_w_link       11938\n",
       "component_w_link      10645\n",
       "class_w_link           6430\n",
       "raw_entry_id              0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_drug_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude = ['component', 'sequence', 'source', 'db_links', 'sequence type', \n",
    "           'dblinks_w_link', 'comment_w_link', 'target_w_link', 'disease', 'disease_w_link',\n",
    "           'remark_w_link', 'image', 'raw_entry_id', 'class', 'original', 'repeat',\n",
    "           'efficacy_w_link','source_w_link','metabolism_w_link','sequence_w_link',\n",
    "            'interaction_w_link','component_w_link','class_w_link', 'interaction','kcf','atom', 'bond',\n",
    "            'bracket', 'remark','metabolism','target']\n",
    "clean_drug_df = raw_drug_df[[col for col in raw_drug_df.columns if col not in exclude]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['entry_id', 'classification', 'name', 'formula', 'exact_mass',\n",
       "       'mol_weight', 'efficacy', 'comment'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_drug_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry_id</th>\n",
       "      <th>classification</th>\n",
       "      <th>name</th>\n",
       "      <th>formula</th>\n",
       "      <th>exact_mass</th>\n",
       "      <th>mol_weight</th>\n",
       "      <th>efficacy</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D00001</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Water (JP18/USP);\\nPurified water (JP18);\\nPur...</td>\n",
       "      <td>H2O</td>\n",
       "      <td>18.0106</td>\n",
       "      <td>18.0153</td>\n",
       "      <td>Pharmaceutic aid (solvent)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D00002</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Nadide (JAN/USAN/INN);\\nNicotinamide adenine d...</td>\n",
       "      <td>C21H28N7O14P2</td>\n",
       "      <td>664.1169</td>\n",
       "      <td>664.4330</td>\n",
       "      <td>Coenzyme</td>\n",
       "      <td>Antagonist (to alcohol and narcotics)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D00003</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Oxygen (JP18/USP)</td>\n",
       "      <td>O2</td>\n",
       "      <td>31.9898</td>\n",
       "      <td>31.9988</td>\n",
       "      <td>Medical gases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D00004</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Carbon dioxide (JP18/USP);\\nCarbon dioxide (TN)</td>\n",
       "      <td>CO2</td>\n",
       "      <td>43.9898</td>\n",
       "      <td>44.0095</td>\n",
       "      <td>Medical gases</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D00005</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Flavin adenine dinucleotide (JAN);\\nAdeflavin ...</td>\n",
       "      <td>C27H33N9O15P2</td>\n",
       "      <td>785.1571</td>\n",
       "      <td>785.5497</td>\n",
       "      <td>Supplement (vitamin B2)</td>\n",
       "      <td>Coenzyme form of vitamin B2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  entry_id classification                                               name  \\\n",
       "0   D00001           Drug  Water (JP18/USP);\\nPurified water (JP18);\\nPur...   \n",
       "1   D00002           Drug  Nadide (JAN/USAN/INN);\\nNicotinamide adenine d...   \n",
       "2   D00003           Drug                                  Oxygen (JP18/USP)   \n",
       "3   D00004           Drug    Carbon dioxide (JP18/USP);\\nCarbon dioxide (TN)   \n",
       "4   D00005           Drug  Flavin adenine dinucleotide (JAN);\\nAdeflavin ...   \n",
       "\n",
       "         formula  exact_mass  mol_weight                    efficacy  \\\n",
       "0            H2O     18.0106     18.0153  Pharmaceutic aid (solvent)   \n",
       "1  C21H28N7O14P2    664.1169    664.4330                    Coenzyme   \n",
       "2             O2     31.9898     31.9988               Medical gases   \n",
       "3            CO2     43.9898     44.0095               Medical gases   \n",
       "4  C27H33N9O15P2    785.1571    785.5497     Supplement (vitamin B2)   \n",
       "\n",
       "                                 comment  \n",
       "0                                    NaN  \n",
       "1  Antagonist (to alcohol and narcotics)  \n",
       "2                                    NaN  \n",
       "3                                    NaN  \n",
       "4            Coenzyme form of vitamin B2  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_drug_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1\n",
    "def row_to_dict(row):\n",
    "    entry_id = row['entry_id']\n",
    "    name = row['name']\n",
    "    efficacy = row['efficacy']\n",
    "    comment = row['comment']\n",
    "    formula = row['formula']\n",
    "    exact_mass = row['exact_mass']\n",
    "    mol_weight = row['mol_weight']\n",
    "    classification = row['classification']\n",
    "    text = f\"name: {name}; formula: {formula}; efficacy: {efficacy}; comment: {comment}\"\n",
    "    output_dict = {\n",
    "            \"id\": entry_id,\n",
    "            \"input\": text,\n",
    "            \"metadata\": {\n",
    "                \"name\": name,\n",
    "                \"text\": text,\n",
    "                \"classification\": classification,\n",
    "                \"formula\": formula,\n",
    "                \"exact mass\": exact_mass,\n",
    "                \"mol_weight\": mol_weight\n",
    "            }\n",
    "        }\n",
    "\n",
    "    return output_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11938\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'id': 'D00001',\n",
       " 'input': 'name: Water (JP18/USP);\\nPurified water (JP18);\\nPurified water in containers (JP18);\\nWater, purified (USP);\\nSterile purified water in containers (JP18);\\nWater for injection (JP18);\\nWater for injection in containers (JP18);\\nSterile water (TN); formula: H2O; efficacy: Pharmaceutic aid (solvent); comment: nan',\n",
       " 'metadata': {'name': 'Water (JP18/USP);\\nPurified water (JP18);\\nPurified water in containers (JP18);\\nWater, purified (USP);\\nSterile purified water in containers (JP18);\\nWater for injection (JP18);\\nWater for injection in containers (JP18);\\nSterile water (TN)',\n",
       "  'text': 'name: Water (JP18/USP);\\nPurified water (JP18);\\nPurified water in containers (JP18);\\nWater, purified (USP);\\nSterile purified water in containers (JP18);\\nWater for injection (JP18);\\nWater for injection in containers (JP18);\\nSterile water (TN); formula: H2O; efficacy: Pharmaceutic aid (solvent); comment: nan',\n",
       "  'classification': 'Drug',\n",
       "  'formula': 'H2O',\n",
       "  'exact mass': 18.0106,\n",
       "  'mol_weight': 18.0153}}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugs_preprocess = convert_dataframe_to_list(clean_drug_df)\n",
    "print(len(drugs_preprocess))\n",
    "drugs_preprocess[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11938, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>input</th>\n",
       "      <th>metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D00001</td>\n",
       "      <td>name: Water (JP18/USP);\\nPurified water (JP18)...</td>\n",
       "      <td>{'name': 'Water (JP18/USP);\n",
       "Purified water (JP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D00002</td>\n",
       "      <td>name: Nadide (JAN/USAN/INN);\\nNicotinamide ade...</td>\n",
       "      <td>{'name': 'Nadide (JAN/USAN/INN);\n",
       "Nicotinamide ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D00003</td>\n",
       "      <td>name: Oxygen (JP18/USP); formula: O2; efficacy...</td>\n",
       "      <td>{'name': 'Oxygen (JP18/USP)', 'text': 'name: O...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D00004</td>\n",
       "      <td>name: Carbon dioxide (JP18/USP);\\nCarbon dioxi...</td>\n",
       "      <td>{'name': 'Carbon dioxide (JP18/USP);\n",
       "Carbon di...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D00005</td>\n",
       "      <td>name: Flavin adenine dinucleotide (JAN);\\nAdef...</td>\n",
       "      <td>{'name': 'Flavin adenine dinucleotide (JAN);\n",
       "A...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                              input  \\\n",
       "0  D00001  name: Water (JP18/USP);\\nPurified water (JP18)...   \n",
       "1  D00002  name: Nadide (JAN/USAN/INN);\\nNicotinamide ade...   \n",
       "2  D00003  name: Oxygen (JP18/USP); formula: O2; efficacy...   \n",
       "3  D00004  name: Carbon dioxide (JP18/USP);\\nCarbon dioxi...   \n",
       "4  D00005  name: Flavin adenine dinucleotide (JAN);\\nAdef...   \n",
       "\n",
       "                                            metadata  \n",
       "0  {'name': 'Water (JP18/USP);\n",
       "Purified water (JP...  \n",
       "1  {'name': 'Nadide (JAN/USAN/INN);\n",
       "Nicotinamide ...  \n",
       "2  {'name': 'Oxygen (JP18/USP)', 'text': 'name: O...  \n",
       "3  {'name': 'Carbon dioxide (JP18/USP);\n",
       "Carbon di...  \n",
       "4  {'name': 'Flavin adenine dinucleotide (JAN);\n",
       "A...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugs_df = pd.DataFrame(drugs_preprocess)\n",
    "preprocessed_datasets.append(drugs_df)\n",
    "print(drugs_df.shape)\n",
    "drugs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kegg Medicus Disease Data Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>NAME</th>\n",
       "      <th>SUBGROUP</th>\n",
       "      <th>SUPERGRP</th>\n",
       "      <th>DESCRIPTION</th>\n",
       "      <th>CATEGORY</th>\n",
       "      <th>NETWORK</th>\n",
       "      <th>GENE</th>\n",
       "      <th>PATHOGEN</th>\n",
       "      <th>ENV_FACTOR</th>\n",
       "      <th>...</th>\n",
       "      <th>DESCRIPTION_link</th>\n",
       "      <th>NETWORK_link</th>\n",
       "      <th>GENE_link</th>\n",
       "      <th>PATHOGEN_link</th>\n",
       "      <th>ENV_FACTOR_link</th>\n",
       "      <th>CARCINOGEN_link</th>\n",
       "      <th>DRUG_link</th>\n",
       "      <th>COMMENT_link</th>\n",
       "      <th>DBLINKS_link</th>\n",
       "      <th>REFERENCE_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H00001</td>\n",
       "      <td>B-cell acute lymphoblastic leukemia;\\n B-cell ...</td>\n",
       "      <td>Philadelphia chromosome positive acute lymphob...</td>\n",
       "      <td>Non-Hodgkin lymphoma [DS:H02418]</td>\n",
       "      <td>Acute lymphocytic leukemia (ALL) is a clonal s...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BCR-ABL (translocation) [HSA:25] [KO:K06619]\\n...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Acute lymphocytic leukemia (ALL) is a clonal s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BCR-ABL (translocation) [HSA:&lt;a href=\"https://...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>*Thorium-232 and its decay products, administe...</td>\n",
       "      <td>Cyclophosphamide [DR:&lt;a href=\"https://www.geno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD-11: 2A70\\n ICD-10: C83.5 C91.0\\n MeSH: &lt;a ...</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H00002</td>\n",
       "      <td>T-cell acute lymphoblastic leukemia;\\n T-cell ...</td>\n",
       "      <td>T-cell lymphoblastic lymphoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acute lymphocytic leukemia (ALL) is a clonal s...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NOTCH1 (mutation) [HSA:4851] [KO:K02599]\\n TAL...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Acute lymphocytic leukemia (ALL) is a clonal s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NOTCH1 (mutation) [HSA:&lt;a href=\"https://www.ge...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>*Thorium-232 and its decay products, administe...</td>\n",
       "      <td>Cyclophosphamide [DR:&lt;a href=\"https://www.geno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD-11: 2A71\\n ICD-10: C83.5 C91.0\\n MeSH: &lt;a ...</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H00003</td>\n",
       "      <td>Acute myeloid leukemia</td>\n",
       "      <td>Acute promyelocytic leukemia [DS:H02542]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acute myeloid leukemia (AML) is a disease that...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>nt06275  Acute myeloid leukemia</td>\n",
       "      <td>PML-RARalpha (translocation) [HSA:5371] [KO:K1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Acute myeloid leukemia (AML) is a disease that...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/network/nt06275\"&gt;...</td>\n",
       "      <td>PML-RARalpha (translocation) [HSA:&lt;a href=\"htt...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benzene [CPD:&lt;a href=\"https://www.genome.jp/db...</td>\n",
       "      <td>Cyclophosphamide [DR:&lt;a href=\"https://www.geno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD-11: 2A60\\n ICD-10: C92.0\\n MeSH: &lt;a href=\"...</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H00004</td>\n",
       "      <td>Chronic myeloid leukemia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chronic myeloid leukemia (CML) is a clonal mye...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>nt06276  Chronic myeloid leukemia</td>\n",
       "      <td>BCR-ABL (translocation) [HSA:613 25] [KO:K0887...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Chronic myeloid leukemia (CML) is a clonal mye...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/network/nt06276\"&gt;...</td>\n",
       "      <td>BCR-ABL (translocation) [HSA:&lt;a href=\"https://...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,3-Butadiene [CPD:&lt;a href=\"https://www.genome...</td>\n",
       "      <td>Busulfan [DR:&lt;a href=\"https://www.genome.jp/db...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD-11: 2A20.0\\n ICD-10: C92.1\\n MeSH: &lt;a href...</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H00005</td>\n",
       "      <td>Chronic lymphocytic leukemia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-Hodgkin lymphoma [DS:H02418]</td>\n",
       "      <td>Chronic lymphocytic leukemia (CLL) is caused b...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bcl-2 (overexpression) [HSA:596] [KO:K02161]\\n...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Chronic lymphocytic leukemia (CLL) is caused b...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bcl-2 (overexpression) [HSA:&lt;a href=\"https://w...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,3-Butadiene [CPD:&lt;a href=\"https://www.genome...</td>\n",
       "      <td>Cyclophosphamide [DR:&lt;a href=\"https://www.geno...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD-11: 2A82.0\\n ICD-10: C91.1\\n MeSH: &lt;a href...</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ENTRY                                               NAME  \\\n",
       "0  H00001  B-cell acute lymphoblastic leukemia;\\n B-cell ...   \n",
       "1  H00002  T-cell acute lymphoblastic leukemia;\\n T-cell ...   \n",
       "2  H00003                             Acute myeloid leukemia   \n",
       "3  H00004                           Chronic myeloid leukemia   \n",
       "4  H00005                       Chronic lymphocytic leukemia   \n",
       "\n",
       "                                            SUBGROUP  \\\n",
       "0  Philadelphia chromosome positive acute lymphob...   \n",
       "1                      T-cell lymphoblastic lymphoma   \n",
       "2           Acute promyelocytic leukemia [DS:H02542]   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "                           SUPERGRP  \\\n",
       "0  Non-Hodgkin lymphoma [DS:H02418]   \n",
       "1                               NaN   \n",
       "2                               NaN   \n",
       "3                               NaN   \n",
       "4  Non-Hodgkin lymphoma [DS:H02418]   \n",
       "\n",
       "                                         DESCRIPTION CATEGORY  \\\n",
       "0  Acute lymphocytic leukemia (ALL) is a clonal s...   Cancer   \n",
       "1  Acute lymphocytic leukemia (ALL) is a clonal s...   Cancer   \n",
       "2  Acute myeloid leukemia (AML) is a disease that...   Cancer   \n",
       "3  Chronic myeloid leukemia (CML) is a clonal mye...   Cancer   \n",
       "4  Chronic lymphocytic leukemia (CLL) is caused b...   Cancer   \n",
       "\n",
       "                             NETWORK  \\\n",
       "0                                NaN   \n",
       "1                                NaN   \n",
       "2    nt06275  Acute myeloid leukemia   \n",
       "3  nt06276  Chronic myeloid leukemia   \n",
       "4                                NaN   \n",
       "\n",
       "                                                GENE PATHOGEN ENV_FACTOR  ...  \\\n",
       "0  BCR-ABL (translocation) [HSA:25] [KO:K06619]\\n...      NaN        NaN  ...   \n",
       "1  NOTCH1 (mutation) [HSA:4851] [KO:K02599]\\n TAL...      NaN        NaN  ...   \n",
       "2  PML-RARalpha (translocation) [HSA:5371] [KO:K1...      NaN        NaN  ...   \n",
       "3  BCR-ABL (translocation) [HSA:613 25] [KO:K0887...      NaN        NaN  ...   \n",
       "4  Bcl-2 (overexpression) [HSA:596] [KO:K02161]\\n...      NaN        NaN  ...   \n",
       "\n",
       "                                    DESCRIPTION_link  \\\n",
       "0  Acute lymphocytic leukemia (ALL) is a clonal s...   \n",
       "1  Acute lymphocytic leukemia (ALL) is a clonal s...   \n",
       "2  Acute myeloid leukemia (AML) is a disease that...   \n",
       "3  Chronic myeloid leukemia (CML) is a clonal mye...   \n",
       "4  Chronic lymphocytic leukemia (CLL) is caused b...   \n",
       "\n",
       "                                        NETWORK_link  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  <a href=\"https://www.kegg.jp/network/nt06275\">...   \n",
       "3  <a href=\"https://www.kegg.jp/network/nt06276\">...   \n",
       "4                                                NaN   \n",
       "\n",
       "                                           GENE_link PATHOGEN_link  \\\n",
       "0  BCR-ABL (translocation) [HSA:<a href=\"https://...           NaN   \n",
       "1  NOTCH1 (mutation) [HSA:<a href=\"https://www.ge...           NaN   \n",
       "2  PML-RARalpha (translocation) [HSA:<a href=\"htt...           NaN   \n",
       "3  BCR-ABL (translocation) [HSA:<a href=\"https://...           NaN   \n",
       "4  Bcl-2 (overexpression) [HSA:<a href=\"https://w...           NaN   \n",
       "\n",
       "  ENV_FACTOR_link                                    CARCINOGEN_link  \\\n",
       "0             NaN  *Thorium-232 and its decay products, administe...   \n",
       "1             NaN  *Thorium-232 and its decay products, administe...   \n",
       "2             NaN  Benzene [CPD:<a href=\"https://www.genome.jp/db...   \n",
       "3             NaN  1,3-Butadiene [CPD:<a href=\"https://www.genome...   \n",
       "4             NaN  1,3-Butadiene [CPD:<a href=\"https://www.genome...   \n",
       "\n",
       "                                           DRUG_link COMMENT_link  \\\n",
       "0  Cyclophosphamide [DR:<a href=\"https://www.geno...          NaN   \n",
       "1  Cyclophosphamide [DR:<a href=\"https://www.geno...          NaN   \n",
       "2  Cyclophosphamide [DR:<a href=\"https://www.geno...          NaN   \n",
       "3  Busulfan [DR:<a href=\"https://www.genome.jp/db...          NaN   \n",
       "4  Cyclophosphamide [DR:<a href=\"https://www.geno...          NaN   \n",
       "\n",
       "                                        DBLINKS_link  \\\n",
       "0  ICD-11: 2A70\\n ICD-10: C83.5 C91.0\\n MeSH: <a ...   \n",
       "1  ICD-11: 2A71\\n ICD-10: C83.5 C91.0\\n MeSH: <a ...   \n",
       "2  ICD-11: 2A60\\n ICD-10: C92.0\\n MeSH: <a href=\"...   \n",
       "3  ICD-11: 2A20.0\\n ICD-10: C92.1\\n MeSH: <a href...   \n",
       "4  ICD-11: 2A82.0\\n ICD-10: C91.1\\n MeSH: <a href...   \n",
       "\n",
       "                                      REFERENCE_link  \n",
       "0  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "1  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "2  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "3  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "4  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_disease_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2577, 28)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_disease_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENTRY                  0\n",
       "NAME                   0\n",
       "SUBGROUP            2168\n",
       "SUPERGRP            1874\n",
       "DESCRIPTION            0\n",
       "CATEGORY               0\n",
       "NETWORK             2400\n",
       "GENE                 516\n",
       "PATHOGEN            2211\n",
       "ENV_FACTOR          2503\n",
       "CARCINOGEN          2537\n",
       "DRUG                2119\n",
       "COMMENT             2145\n",
       "DBLINKS               35\n",
       "REFERENCE              3\n",
       "ENTRY_link             0\n",
       "SUBGROUP_link       2168\n",
       "SUPERGRP_link       1874\n",
       "DESCRIPTION_link       0\n",
       "NETWORK_link        2400\n",
       "GENE_link            516\n",
       "PATHOGEN_link       2211\n",
       "ENV_FACTOR_link     2503\n",
       "CARCINOGEN_link     2537\n",
       "DRUG_link           2119\n",
       "COMMENT_link        2145\n",
       "DBLINKS_link          35\n",
       "REFERENCE_link         3\n",
       "dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_disease_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>NAME</th>\n",
       "      <th>DESCRIPTION</th>\n",
       "      <th>CATEGORY</th>\n",
       "      <th>GENE</th>\n",
       "      <th>PATHOGEN</th>\n",
       "      <th>ENV_FACTOR</th>\n",
       "      <th>CARCINOGEN</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>COMMENT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H00001</td>\n",
       "      <td>B-cell acute lymphoblastic leukemia;\\n B-cell ...</td>\n",
       "      <td>Acute lymphocytic leukemia (ALL) is a clonal s...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>BCR-ABL (translocation) [HSA:25] [KO:K06619]\\n...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>*Thorium-232 and its decay products, administe...</td>\n",
       "      <td>Cyclophosphamide [DR:D00287]\\n Methotrexate [D...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H00002</td>\n",
       "      <td>T-cell acute lymphoblastic leukemia;\\n T-cell ...</td>\n",
       "      <td>Acute lymphocytic leukemia (ALL) is a clonal s...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>NOTCH1 (mutation) [HSA:4851] [KO:K02599]\\n TAL...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>*Thorium-232 and its decay products, administe...</td>\n",
       "      <td>Cyclophosphamide [DR:D00287]\\n Methotrexate [D...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H00003</td>\n",
       "      <td>Acute myeloid leukemia</td>\n",
       "      <td>Acute myeloid leukemia (AML) is a disease that...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>PML-RARalpha (translocation) [HSA:5371] [KO:K1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benzene [CPD:C01407]\\n 1,4-Butanediol dimethan...</td>\n",
       "      <td>Cyclophosphamide [DR:D00287]\\n Thioguanine [DR...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H00004</td>\n",
       "      <td>Chronic myeloid leukemia</td>\n",
       "      <td>Chronic myeloid leukemia (CML) is a clonal mye...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>BCR-ABL (translocation) [HSA:613 25] [KO:K0887...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,3-Butadiene [CPD:C16450]\\n Rubber industry</td>\n",
       "      <td>Busulfan [DR:D00248]\\n Thioguanine [DR:D06109]...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H00005</td>\n",
       "      <td>Chronic lymphocytic leukemia</td>\n",
       "      <td>Chronic lymphocytic leukemia (CLL) is caused b...</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Bcl-2 (overexpression) [HSA:596] [KO:K02161]\\n...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1,3-Butadiene [CPD:C16450]\\n Ethylene oxide [C...</td>\n",
       "      <td>Cyclophosphamide [DR:D00287]\\n Chlorambucil [D...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ENTRY                                               NAME  \\\n",
       "0  H00001  B-cell acute lymphoblastic leukemia;\\n B-cell ...   \n",
       "1  H00002  T-cell acute lymphoblastic leukemia;\\n T-cell ...   \n",
       "2  H00003                             Acute myeloid leukemia   \n",
       "3  H00004                           Chronic myeloid leukemia   \n",
       "4  H00005                       Chronic lymphocytic leukemia   \n",
       "\n",
       "                                         DESCRIPTION CATEGORY  \\\n",
       "0  Acute lymphocytic leukemia (ALL) is a clonal s...   Cancer   \n",
       "1  Acute lymphocytic leukemia (ALL) is a clonal s...   Cancer   \n",
       "2  Acute myeloid leukemia (AML) is a disease that...   Cancer   \n",
       "3  Chronic myeloid leukemia (CML) is a clonal mye...   Cancer   \n",
       "4  Chronic lymphocytic leukemia (CLL) is caused b...   Cancer   \n",
       "\n",
       "                                                GENE PATHOGEN ENV_FACTOR  \\\n",
       "0  BCR-ABL (translocation) [HSA:25] [KO:K06619]\\n...      NaN        NaN   \n",
       "1  NOTCH1 (mutation) [HSA:4851] [KO:K02599]\\n TAL...      NaN        NaN   \n",
       "2  PML-RARalpha (translocation) [HSA:5371] [KO:K1...      NaN        NaN   \n",
       "3  BCR-ABL (translocation) [HSA:613 25] [KO:K0887...      NaN        NaN   \n",
       "4  Bcl-2 (overexpression) [HSA:596] [KO:K02161]\\n...      NaN        NaN   \n",
       "\n",
       "                                          CARCINOGEN  \\\n",
       "0  *Thorium-232 and its decay products, administe...   \n",
       "1  *Thorium-232 and its decay products, administe...   \n",
       "2  Benzene [CPD:C01407]\\n 1,4-Butanediol dimethan...   \n",
       "3       1,3-Butadiene [CPD:C16450]\\n Rubber industry   \n",
       "4  1,3-Butadiene [CPD:C16450]\\n Ethylene oxide [C...   \n",
       "\n",
       "                                                DRUG COMMENT  \n",
       "0  Cyclophosphamide [DR:D00287]\\n Methotrexate [D...     NaN  \n",
       "1  Cyclophosphamide [DR:D00287]\\n Methotrexate [D...     NaN  \n",
       "2  Cyclophosphamide [DR:D00287]\\n Thioguanine [DR...     NaN  \n",
       "3  Busulfan [DR:D00248]\\n Thioguanine [DR:D06109]...     NaN  \n",
       "4  Cyclophosphamide [DR:D00287]\\n Chlorambucil [D...     NaN  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exclude = ['SUBGROUP', 'SUPERGRP', 'NETWORK', 'ENTRY_link',\n",
    "           'SUBGROUP_link', 'SUPERGRP_link', 'DESCRIPTION_link',\n",
    "           'NETWORK_link', 'GENE_link', 'PATHOGEN_link', 'ENV_FACTOR_link',\n",
    "           'CARCINOGEN_link', 'DRUG_link', 'COMMENT_link','DBLINKS_link',\n",
    "           'REFERENCE_link', 'DBLINKS', 'REFERENCE']\n",
    "clean_disease_df = raw_disease_df[[col for col in raw_disease_df.columns if col not in exclude]]\n",
    "clean_disease_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2\n",
    "def row_to_dict(row):\n",
    "    entry_id = row['ENTRY']\n",
    "    name = row['NAME']\n",
    "    description = row['DESCRIPTION']\n",
    "    comment = row['COMMENT']\n",
    "    category = row['CATEGORY']\n",
    "    gene = row['GENE']\n",
    "    pathogen = row['PATHOGEN']\n",
    "    env_factor = row['ENV_FACTOR']\n",
    "    carcinogen = row['CARCINOGEN']\n",
    "    drug = row['DRUG']\n",
    "    text = f\"name: {name}; category: {category}; description: {description}; drug: {drug}\"\n",
    "\n",
    "    output_dict = {\n",
    "            \"id\": entry_id,\n",
    "            \"input\": text,\n",
    "            \"metadata\": {\n",
    "                \"name\": name,\n",
    "                \"text\": text,\n",
    "                \"gene\": gene,\n",
    "                \"pathogen\": pathogen,\n",
    "                \"env_factor\": env_factor,\n",
    "                \"carcinogen\": carcinogen,\n",
    "                \"drug\": drug,\n",
    "                \"comment\": comment,\n",
    "            }\n",
    "        }\n",
    "\n",
    "    return output_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2577\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'id': 'H00002',\n",
       " 'input': 'name: T-cell acute lymphoblastic leukemia;\\n T-cell acute lymphocytic leukemia; category: Cancer; description: Acute lymphocytic leukemia (ALL) is a clonal stem cell malignancy of excessive lymphoblast proliferation. It is now understood that ALL and lymphoblastic lymphoma are the same disease entities at the morphologic and immunophenotypic levels and classified as either B- and T-cell lymphoblastic leukemia/lymphoma (B-ALL and T-ALL). T-ALL comprises 15% of paediatric and 25% of adult ALL cases. T cell transformation is a multi-step process in which different genetic alterations cooperate to alter the normal mechanisms that control cell growth, proliferation, survival, and differentiation during thymocyte development. In this context, constitutive activation of NOTCH1 signaling is the most prominent oncogenic pathway in T cell transformation. In addition, T-ALLs characteristically show the translocation and aberrant expression of transcription factor oncogenes. These oncogenic transcription factors include T-cell leukaemia homeobox protein 1 (TLX1 also known as HOX11), TLX3 (HOX11L2), LYL1, TAL1 and MLL.; drug: Cyclophosphamide [DR:D00287]\\n Methotrexate [DR:D00142]\\n Mercaptopurine [DR:D00161]\\n Clofarabine [DR:D03546]\\n Nelarabine [DR:D05134]\\n Vincristine sulfate [DR:D02197] (Philadelphia chromosome-negative)\\n Teniposide [DR:D02698]\\n Doxorubicin hydrochloride [DR:D01275]\\n Daunorubicin hydrochloride [DR:D01264]\\n Mitoxantrone hydrochloride [DR:D02166]\\n Asparaginase [DR:D02997]\\n Pegaspargase [DR:D05387]\\n Enocitabine [DR:D01633]\\n Calaspargase pegol [DR:D10096]',\n",
       " 'metadata': {'name': 'T-cell acute lymphoblastic leukemia;\\n T-cell acute lymphocytic leukemia',\n",
       "  'text': 'name: T-cell acute lymphoblastic leukemia;\\n T-cell acute lymphocytic leukemia; category: Cancer; description: Acute lymphocytic leukemia (ALL) is a clonal stem cell malignancy of excessive lymphoblast proliferation. It is now understood that ALL and lymphoblastic lymphoma are the same disease entities at the morphologic and immunophenotypic levels and classified as either B- and T-cell lymphoblastic leukemia/lymphoma (B-ALL and T-ALL). T-ALL comprises 15% of paediatric and 25% of adult ALL cases. T cell transformation is a multi-step process in which different genetic alterations cooperate to alter the normal mechanisms that control cell growth, proliferation, survival, and differentiation during thymocyte development. In this context, constitutive activation of NOTCH1 signaling is the most prominent oncogenic pathway in T cell transformation. In addition, T-ALLs characteristically show the translocation and aberrant expression of transcription factor oncogenes. These oncogenic transcription factors include T-cell leukaemia homeobox protein 1 (TLX1 also known as HOX11), TLX3 (HOX11L2), LYL1, TAL1 and MLL.; drug: Cyclophosphamide [DR:D00287]\\n Methotrexate [DR:D00142]\\n Mercaptopurine [DR:D00161]\\n Clofarabine [DR:D03546]\\n Nelarabine [DR:D05134]\\n Vincristine sulfate [DR:D02197] (Philadelphia chromosome-negative)\\n Teniposide [DR:D02698]\\n Doxorubicin hydrochloride [DR:D01275]\\n Daunorubicin hydrochloride [DR:D01264]\\n Mitoxantrone hydrochloride [DR:D02166]\\n Asparaginase [DR:D02997]\\n Pegaspargase [DR:D05387]\\n Enocitabine [DR:D01633]\\n Calaspargase pegol [DR:D10096]',\n",
       "  'gene': 'NOTCH1 (mutation) [HSA:4851] [KO:K02599]\\n TAL1 (overexpression) [HSA:6886] [KO:K09068]\\n TAL2 [HSA:6887] [KO:K09068]\\n LYL1 (expression) [HSA:4066] [KO:K15604]\\n MLL-ENL (translocation) [HSA:4297] [KO:K09186]\\n HOX11 (translocation) [HSA:3195] [KO:K09340]\\n MYC (translocation) [HSA:4609] [KO:K04377]\\n LMO2 (translocation) [HSA:4005] [KO:K15612]\\n HOX11L2 (translocation) [HSA:30012] [KO:K15607]\\n PICALM-MLLT10 (translocation) [HSA:8028] [KO:K23588]',\n",
       "  'pathogen': nan,\n",
       "  'env_factor': nan,\n",
       "  'carcinogen': '*Thorium-232 and its decay products, administered intravenously as a colloidal dispersion of thorium232 dioxide\\n *: Although this is a risk factor for ALL in general, the strength of association with different histological tumor types is not well understood.',\n",
       "  'drug': 'Cyclophosphamide [DR:D00287]\\n Methotrexate [DR:D00142]\\n Mercaptopurine [DR:D00161]\\n Clofarabine [DR:D03546]\\n Nelarabine [DR:D05134]\\n Vincristine sulfate [DR:D02197] (Philadelphia chromosome-negative)\\n Teniposide [DR:D02698]\\n Doxorubicin hydrochloride [DR:D01275]\\n Daunorubicin hydrochloride [DR:D01264]\\n Mitoxantrone hydrochloride [DR:D02166]\\n Asparaginase [DR:D02997]\\n Pegaspargase [DR:D05387]\\n Enocitabine [DR:D01633]\\n Calaspargase pegol [DR:D10096]',\n",
       "  'comment': nan}}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_preprocess = convert_dataframe_to_list(clean_disease_df)\n",
    "print(len(disease_preprocess))\n",
    "disease_preprocess[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2577, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>input</th>\n",
       "      <th>metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H00001</td>\n",
       "      <td>name: B-cell acute lymphoblastic leukemia;\\n B...</td>\n",
       "      <td>{'name': 'B-cell acute lymphoblastic leukemia;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H00002</td>\n",
       "      <td>name: T-cell acute lymphoblastic leukemia;\\n T...</td>\n",
       "      <td>{'name': 'T-cell acute lymphoblastic leukemia;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H00003</td>\n",
       "      <td>name: Acute myeloid leukemia; category: Cancer...</td>\n",
       "      <td>{'name': 'Acute myeloid leukemia', 'text': 'na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H00004</td>\n",
       "      <td>name: Chronic myeloid leukemia; category: Canc...</td>\n",
       "      <td>{'name': 'Chronic myeloid leukemia', 'text': '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H00005</td>\n",
       "      <td>name: Chronic lymphocytic leukemia; category: ...</td>\n",
       "      <td>{'name': 'Chronic lymphocytic leukemia', 'text...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                              input  \\\n",
       "0  H00001  name: B-cell acute lymphoblastic leukemia;\\n B...   \n",
       "1  H00002  name: T-cell acute lymphoblastic leukemia;\\n T...   \n",
       "2  H00003  name: Acute myeloid leukemia; category: Cancer...   \n",
       "3  H00004  name: Chronic myeloid leukemia; category: Canc...   \n",
       "4  H00005  name: Chronic lymphocytic leukemia; category: ...   \n",
       "\n",
       "                                            metadata  \n",
       "0  {'name': 'B-cell acute lymphoblastic leukemia;...  \n",
       "1  {'name': 'T-cell acute lymphoblastic leukemia;...  \n",
       "2  {'name': 'Acute myeloid leukemia', 'text': 'na...  \n",
       "3  {'name': 'Chronic myeloid leukemia', 'text': '...  \n",
       "4  {'name': 'Chronic lymphocytic leukemia', 'text...  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_df = pd.DataFrame(disease_preprocess)\n",
    "preprocessed_datasets.append(disease_df)\n",
    "print(disease_df.shape)\n",
    "disease_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kegg Medicus D-Group Data Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2398, 13)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_dgroup_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>CLASSIFICATION</th>\n",
       "      <th>NAME</th>\n",
       "      <th>STEM</th>\n",
       "      <th>IMAGE</th>\n",
       "      <th>MEMBER</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>REMARK</th>\n",
       "      <th>COMMENT</th>\n",
       "      <th>ENTRY_link</th>\n",
       "      <th>MEMBER_link</th>\n",
       "      <th>CLASS_link</th>\n",
       "      <th>REMARK_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DG00001</td>\n",
       "      <td>DGroup</td>\n",
       "      <td>Chlorhexidine\\n TYPE        Chemical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D07668  Chlorhexidine (INN)\\n D00858  Chlorhex...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATC code: A01AB03 B05CA02 D08AC02 D09AA12 R02A...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATC code: &lt;a href=\"https://www.kegg.jp/kegg-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DG00002</td>\n",
       "      <td>DGroup</td>\n",
       "      <td>Oxyquinoline\\n TYPE        Chemical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D05321  Oxyquinoline (USAN)\\n D02414  Oxyquino...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATC code: A01AB07 D08AH03 G01AC30 R02AA14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATC code: &lt;a href=\"https://www.kegg.jp/kegg-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DG00003</td>\n",
       "      <td>DGroup</td>\n",
       "      <td>Neomycin\\n ABBR      NEO\\n TYPE        Chemical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D08260  Neomycin (INN)\\n D01618  Neomycin sulf...</td>\n",
       "      <td>Antibacterial\\n DG01447  Aminoglycoside antibi...</td>\n",
       "      <td>ATC code: A01AB08 A07AA01 B05CA09 D06AX04 J01G...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>Antibacterial\\n &lt;a href=\"https://www.kegg.jp/d...</td>\n",
       "      <td>ATC code: &lt;a href=\"https://www.kegg.jp/kegg-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DG00004</td>\n",
       "      <td>DGroup</td>\n",
       "      <td>Miconazole\\n TYPE        Chemical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D00416  Miconazole (JP18/USP/INN)  &lt;JP/US&gt;\\n D...</td>\n",
       "      <td>Antifungal\\n DG01883  Imidazole antifungal\\n M...</td>\n",
       "      <td>ATC code: A01AB09 A07AC01 D01AC02 G01AF04 J02A...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>Antifungal\\n &lt;a href=\"https://www.kegg.jp/dbge...</td>\n",
       "      <td>ATC code: &lt;a href=\"https://www.kegg.jp/kegg-bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DG00005</td>\n",
       "      <td>DGroup</td>\n",
       "      <td>Tetracycline\\n ABBR      TET\\n TYPE        Che...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D00201  Tetracycline (JAN/USP/INN)\\n D02122  T...</td>\n",
       "      <td>Antibacterial\\n DG01197  Tetracycline antibiotic</td>\n",
       "      <td>ATC code: A01AB13 D06AA04 J01AA07 S01AA09 S02A...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>Antibacterial\\n &lt;a href=\"https://www.kegg.jp/d...</td>\n",
       "      <td>ATC code: &lt;a href=\"https://www.kegg.jp/kegg-bi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ENTRY CLASSIFICATION                                               NAME  \\\n",
       "0  DG00001         DGroup               Chlorhexidine\\n TYPE        Chemical   \n",
       "1  DG00002         DGroup                Oxyquinoline\\n TYPE        Chemical   \n",
       "2  DG00003         DGroup    Neomycin\\n ABBR      NEO\\n TYPE        Chemical   \n",
       "3  DG00004         DGroup                  Miconazole\\n TYPE        Chemical   \n",
       "4  DG00005         DGroup  Tetracycline\\n ABBR      TET\\n TYPE        Che...   \n",
       "\n",
       "  STEM  IMAGE                                             MEMBER  \\\n",
       "0  NaN    NaN  D07668  Chlorhexidine (INN)\\n D00858  Chlorhex...   \n",
       "1  NaN    NaN  D05321  Oxyquinoline (USAN)\\n D02414  Oxyquino...   \n",
       "2  NaN    NaN  D08260  Neomycin (INN)\\n D01618  Neomycin sulf...   \n",
       "3  NaN    NaN  D00416  Miconazole (JP18/USP/INN)  <JP/US>\\n D...   \n",
       "4  NaN    NaN  D00201  Tetracycline (JAN/USP/INN)\\n D02122  T...   \n",
       "\n",
       "                                               CLASS  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  Antibacterial\\n DG01447  Aminoglycoside antibi...   \n",
       "3  Antifungal\\n DG01883  Imidazole antifungal\\n M...   \n",
       "4   Antibacterial\\n DG01197  Tetracycline antibiotic   \n",
       "\n",
       "                                              REMARK COMMENT  \\\n",
       "0  ATC code: A01AB03 B05CA02 D08AC02 D09AA12 R02A...     NaN   \n",
       "1          ATC code: A01AB07 D08AH03 G01AC30 R02AA14     NaN   \n",
       "2  ATC code: A01AB08 A07AA01 B05CA09 D06AX04 J01G...     NaN   \n",
       "3  ATC code: A01AB09 A07AC01 D01AC02 G01AF04 J02A...     NaN   \n",
       "4  ATC code: A01AB13 D06AA04 J01AA07 S01AA09 S02A...     NaN   \n",
       "\n",
       "                                          ENTRY_link  \\\n",
       "0  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "1  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "2  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "3  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "4  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "\n",
       "                                         MEMBER_link  \\\n",
       "0  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "1  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "2  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "3  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "4  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "\n",
       "                                          CLASS_link  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  Antibacterial\\n <a href=\"https://www.kegg.jp/d...   \n",
       "3  Antifungal\\n <a href=\"https://www.kegg.jp/dbge...   \n",
       "4  Antibacterial\\n <a href=\"https://www.kegg.jp/d...   \n",
       "\n",
       "                                         REMARK_link  \n",
       "0  ATC code: <a href=\"https://www.kegg.jp/kegg-bi...  \n",
       "1  ATC code: <a href=\"https://www.kegg.jp/kegg-bi...  \n",
       "2  ATC code: <a href=\"https://www.kegg.jp/kegg-bi...  \n",
       "3  ATC code: <a href=\"https://www.kegg.jp/kegg-bi...  \n",
       "4  ATC code: <a href=\"https://www.kegg.jp/kegg-bi...  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_dgroup_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude = ['CLASSIFICATION','STEM','IMAGE','COMMENT','CLASS_link','ENTRY_link','REMARK_link','MEMBER_link']\n",
    "clean_dgroup_df = raw_dgroup_df[[col for col in raw_dgroup_df.columns if col not in exclude]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENTRY                0\n",
       "CLASSIFICATION       0\n",
       "NAME                 0\n",
       "STEM              2232\n",
       "IMAGE             2398\n",
       "MEMBER               0\n",
       "CLASS              570\n",
       "REMARK             966\n",
       "COMMENT           2387\n",
       "ENTRY_link           0\n",
       "MEMBER_link          0\n",
       "CLASS_link         570\n",
       "REMARK_link        966\n",
       "dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_dgroup_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>NAME</th>\n",
       "      <th>MEMBER</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>REMARK</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DG00001</td>\n",
       "      <td>Chlorhexidine\\n TYPE        Chemical</td>\n",
       "      <td>D07668  Chlorhexidine (INN)\\n D00858  Chlorhex...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATC code: A01AB03 B05CA02 D08AC02 D09AA12 R02A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DG00002</td>\n",
       "      <td>Oxyquinoline\\n TYPE        Chemical</td>\n",
       "      <td>D05321  Oxyquinoline (USAN)\\n D02414  Oxyquino...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATC code: A01AB07 D08AH03 G01AC30 R02AA14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DG00003</td>\n",
       "      <td>Neomycin\\n ABBR      NEO\\n TYPE        Chemical</td>\n",
       "      <td>D08260  Neomycin (INN)\\n D01618  Neomycin sulf...</td>\n",
       "      <td>Antibacterial\\n DG01447  Aminoglycoside antibi...</td>\n",
       "      <td>ATC code: A01AB08 A07AA01 B05CA09 D06AX04 J01G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DG00004</td>\n",
       "      <td>Miconazole\\n TYPE        Chemical</td>\n",
       "      <td>D00416  Miconazole (JP18/USP/INN)  &lt;JP/US&gt;\\n D...</td>\n",
       "      <td>Antifungal\\n DG01883  Imidazole antifungal\\n M...</td>\n",
       "      <td>ATC code: A01AB09 A07AC01 D01AC02 G01AF04 J02A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DG00005</td>\n",
       "      <td>Tetracycline\\n ABBR      TET\\n TYPE        Che...</td>\n",
       "      <td>D00201  Tetracycline (JAN/USP/INN)\\n D02122  T...</td>\n",
       "      <td>Antibacterial\\n DG01197  Tetracycline antibiotic</td>\n",
       "      <td>ATC code: A01AB13 D06AA04 J01AA07 S01AA09 S02A...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ENTRY                                               NAME  \\\n",
       "0  DG00001               Chlorhexidine\\n TYPE        Chemical   \n",
       "1  DG00002                Oxyquinoline\\n TYPE        Chemical   \n",
       "2  DG00003    Neomycin\\n ABBR      NEO\\n TYPE        Chemical   \n",
       "3  DG00004                  Miconazole\\n TYPE        Chemical   \n",
       "4  DG00005  Tetracycline\\n ABBR      TET\\n TYPE        Che...   \n",
       "\n",
       "                                              MEMBER  \\\n",
       "0  D07668  Chlorhexidine (INN)\\n D00858  Chlorhex...   \n",
       "1  D05321  Oxyquinoline (USAN)\\n D02414  Oxyquino...   \n",
       "2  D08260  Neomycin (INN)\\n D01618  Neomycin sulf...   \n",
       "3  D00416  Miconazole (JP18/USP/INN)  <JP/US>\\n D...   \n",
       "4  D00201  Tetracycline (JAN/USP/INN)\\n D02122  T...   \n",
       "\n",
       "                                               CLASS  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  Antibacterial\\n DG01447  Aminoglycoside antibi...   \n",
       "3  Antifungal\\n DG01883  Imidazole antifungal\\n M...   \n",
       "4   Antibacterial\\n DG01197  Tetracycline antibiotic   \n",
       "\n",
       "                                              REMARK  \n",
       "0  ATC code: A01AB03 B05CA02 D08AC02 D09AA12 R02A...  \n",
       "1          ATC code: A01AB07 D08AH03 G01AC30 R02AA14  \n",
       "2  ATC code: A01AB08 A07AA01 B05CA09 D06AX04 J01G...  \n",
       "3  ATC code: A01AB09 A07AC01 D01AC02 G01AF04 J02A...  \n",
       "4  ATC code: A01AB13 D06AA04 J01AA07 S01AA09 S02A...  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_dgroup_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3\n",
    "def row_to_dict(row):\n",
    "    entry_id = row['ENTRY']\n",
    "    name = row['NAME']\n",
    "    class_item = row['CLASS']\n",
    "    remark = row['REMARK']\n",
    "    member = row['MEMBER']\n",
    "    text = f\"name: {name}; member: {member}; class: {class_item};  remark: {remark}\"\n",
    "\n",
    "    output_dict = {\n",
    "            \"id\": entry_id,\n",
    "            \"input\": text,\n",
    "            \"metadata\": {\n",
    "                \"text\": text,\n",
    "                \"name\": name,\n",
    "                \"class\": class_item\n",
    "            }\n",
    "        }\n",
    "\n",
    "    return output_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2398\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'id': 'DG00001',\n",
       " 'input': 'name: Chlorhexidine\\n TYPE        Chemical; member: D07668  Chlorhexidine (INN)\\n D00858  Chlorhexidine gluconate (JP18/USP)  <JP/US>\\n D01345  Chlorhexidine hydrochloride (JP18/USP)\\n D03463  Chlorhexidine phosphanilate (USAN)\\n D07669  Chlorhexidine acetate (USP); class: nan;  remark: ATC code: A01AB03 B05CA02 D08AC02 D09AA12 R02AA05 S01AX09 S02AA09 S03AA04',\n",
       " 'metadata': {'text': 'name: Chlorhexidine\\n TYPE        Chemical; member: D07668  Chlorhexidine (INN)\\n D00858  Chlorhexidine gluconate (JP18/USP)  <JP/US>\\n D01345  Chlorhexidine hydrochloride (JP18/USP)\\n D03463  Chlorhexidine phosphanilate (USAN)\\n D07669  Chlorhexidine acetate (USP); class: nan;  remark: ATC code: A01AB03 B05CA02 D08AC02 D09AA12 R02AA05 S01AX09 S02AA09 S03AA04',\n",
       "  'name': 'Chlorhexidine\\n TYPE        Chemical',\n",
       "  'class': nan}}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgroup_preprocess = convert_dataframe_to_list(clean_dgroup_df)\n",
    "print(len(dgroup_preprocess))\n",
    "dgroup_preprocess[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2398, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>input</th>\n",
       "      <th>metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DG00001</td>\n",
       "      <td>name: Chlorhexidine\\n TYPE        Chemical; me...</td>\n",
       "      <td>{'text': 'name: Chlorhexidine\n",
       " TYPE        Che...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DG00002</td>\n",
       "      <td>name: Oxyquinoline\\n TYPE        Chemical; mem...</td>\n",
       "      <td>{'text': 'name: Oxyquinoline\n",
       " TYPE        Chem...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DG00003</td>\n",
       "      <td>name: Neomycin\\n ABBR      NEO\\n TYPE        C...</td>\n",
       "      <td>{'text': 'name: Neomycin\n",
       " ABBR      NEO\n",
       " TYPE ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DG00004</td>\n",
       "      <td>name: Miconazole\\n TYPE        Chemical; membe...</td>\n",
       "      <td>{'text': 'name: Miconazole\n",
       " TYPE        Chemic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DG00005</td>\n",
       "      <td>name: Tetracycline\\n ABBR      TET\\n TYPE     ...</td>\n",
       "      <td>{'text': 'name: Tetracycline\n",
       " ABBR      TET\n",
       " T...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id                                              input  \\\n",
       "0  DG00001  name: Chlorhexidine\\n TYPE        Chemical; me...   \n",
       "1  DG00002  name: Oxyquinoline\\n TYPE        Chemical; mem...   \n",
       "2  DG00003  name: Neomycin\\n ABBR      NEO\\n TYPE        C...   \n",
       "3  DG00004  name: Miconazole\\n TYPE        Chemical; membe...   \n",
       "4  DG00005  name: Tetracycline\\n ABBR      TET\\n TYPE     ...   \n",
       "\n",
       "                                            metadata  \n",
       "0  {'text': 'name: Chlorhexidine\n",
       " TYPE        Che...  \n",
       "1  {'text': 'name: Oxyquinoline\n",
       " TYPE        Chem...  \n",
       "2  {'text': 'name: Neomycin\n",
       " ABBR      NEO\n",
       " TYPE ...  \n",
       "3  {'text': 'name: Miconazole\n",
       " TYPE        Chemic...  \n",
       "4  {'text': 'name: Tetracycline\n",
       " ABBR      TET\n",
       " T...  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgroup_df = pd.DataFrame(dgroup_preprocess)\n",
    "preprocessed_datasets.append(dgroup_df)\n",
    "print(dgroup_df.shape)\n",
    "dgroup_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kegg Medicus Environ Data Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E number</th>\n",
       "      <th>Name</th>\n",
       "      <th>Category</th>\n",
       "      <th>Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Remark</th>\n",
       "      <th>Comment</th>\n",
       "      <th>Other DBs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E00001</td>\n",
       "      <td>Coptis rhizome (JP17)</td>\n",
       "      <td>Crude drug</td>\n",
       "      <td>Berberine [CPD:C00757], Palmatine [CPD:C05315]...</td>\n",
       "      <td>Coptis japonica [TAX:3442], Coptis chinensis [...</td>\n",
       "      <td>Same as: D00092</td>\n",
       "      <td>Ranunculaceae (buttercup family) Coptis rhizom...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E00002</td>\n",
       "      <td>Evodia fruit (JP17)</td>\n",
       "      <td>Crude drug</td>\n",
       "      <td>Evodiamine [CPD:C09187], Rutecarpine [CPD:C092...</td>\n",
       "      <td>Evodia rutaecarpa [TAX:354523], Evodia officin...</td>\n",
       "      <td>Same as: D00173</td>\n",
       "      <td>Rutaceae (rue family) Evodia fruit\\r\\nMajor co...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E00003</td>\n",
       "      <td>Saffron (JP17)</td>\n",
       "      <td>Crude drug; Medicinal herb</td>\n",
       "      <td>Crocin [CPD:C08589], Picrocrocin [CPD:C17055],...</td>\n",
       "      <td>Crocus sativus [TAX:82528]</td>\n",
       "      <td>Same as: D01030</td>\n",
       "      <td>Iridaceae (Iris family) Saffron chapiter\\r\\nMa...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E00004</td>\n",
       "      <td>Agar (JP17/NF)</td>\n",
       "      <td>Crude drug</td>\n",
       "      <td>Agarose [CPD:C01399]</td>\n",
       "      <td>Gelidium amansii [TAX:2812]</td>\n",
       "      <td>Same as: D01032</td>\n",
       "      <td>Gelidiaceae Gelidium amansii mucous (freeze dr...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E00005</td>\n",
       "      <td>Tragacanth (JP17/NF)</td>\n",
       "      <td>Crude drug</td>\n",
       "      <td>D-Galacturonic acid [CPD:C00333], Tragacanthic...</td>\n",
       "      <td>Astragalus gummifer [TAX:339493], Leguminosae ...</td>\n",
       "      <td>Same as: D01033</td>\n",
       "      <td>Fabaceae (pea family) Tragacanth stem exudatio...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  E number                   Name                    Category  \\\n",
       "0   E00001  Coptis rhizome (JP17)                  Crude drug   \n",
       "1   E00002    Evodia fruit (JP17)                  Crude drug   \n",
       "2   E00003         Saffron (JP17)  Crude drug; Medicinal herb   \n",
       "3   E00004         Agar (JP17/NF)                  Crude drug   \n",
       "4   E00005   Tragacanth (JP17/NF)                  Crude drug   \n",
       "\n",
       "                                           Component  \\\n",
       "0  Berberine [CPD:C00757], Palmatine [CPD:C05315]...   \n",
       "1  Evodiamine [CPD:C09187], Rutecarpine [CPD:C092...   \n",
       "2  Crocin [CPD:C08589], Picrocrocin [CPD:C17055],...   \n",
       "3                               Agarose [CPD:C01399]   \n",
       "4  D-Galacturonic acid [CPD:C00333], Tragacanthic...   \n",
       "\n",
       "                                              Source           Remark  \\\n",
       "0  Coptis japonica [TAX:3442], Coptis chinensis [...  Same as: D00092   \n",
       "1  Evodia rutaecarpa [TAX:354523], Evodia officin...  Same as: D00173   \n",
       "2                         Crocus sativus [TAX:82528]  Same as: D01030   \n",
       "3                        Gelidium amansii [TAX:2812]  Same as: D01032   \n",
       "4  Astragalus gummifer [TAX:339493], Leguminosae ...  Same as: D01033   \n",
       "\n",
       "                                             Comment Other DBs  \n",
       "0  Ranunculaceae (buttercup family) Coptis rhizom...       NaN  \n",
       "1  Rutaceae (rue family) Evodia fruit\\r\\nMajor co...       NaN  \n",
       "2  Iridaceae (Iris family) Saffron chapiter\\r\\nMa...       NaN  \n",
       "3  Gelidiaceae Gelidium amansii mucous (freeze dr...       NaN  \n",
       "4  Fabaceae (pea family) Tragacanth stem exudatio...       NaN  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_environ_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(850, 8)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_environ_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude = ['Remark', 'Other DBs']\n",
    "clean_environ_df = raw_environ_df[[col for col in raw_environ_df.columns if col not in exclude]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "E number       0\n",
       "Name           0\n",
       "Category      12\n",
       "Component    306\n",
       "Source        22\n",
       "Remark       556\n",
       "Comment        9\n",
       "Other DBs    757\n",
       "dtype: int64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_environ_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E number</th>\n",
       "      <th>Name</th>\n",
       "      <th>Category</th>\n",
       "      <th>Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E00001</td>\n",
       "      <td>Coptis rhizome (JP17)</td>\n",
       "      <td>Crude drug</td>\n",
       "      <td>Berberine [CPD:C00757], Palmatine [CPD:C05315]...</td>\n",
       "      <td>Coptis japonica [TAX:3442], Coptis chinensis [...</td>\n",
       "      <td>Ranunculaceae (buttercup family) Coptis rhizom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E00002</td>\n",
       "      <td>Evodia fruit (JP17)</td>\n",
       "      <td>Crude drug</td>\n",
       "      <td>Evodiamine [CPD:C09187], Rutecarpine [CPD:C092...</td>\n",
       "      <td>Evodia rutaecarpa [TAX:354523], Evodia officin...</td>\n",
       "      <td>Rutaceae (rue family) Evodia fruit\\r\\nMajor co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E00003</td>\n",
       "      <td>Saffron (JP17)</td>\n",
       "      <td>Crude drug; Medicinal herb</td>\n",
       "      <td>Crocin [CPD:C08589], Picrocrocin [CPD:C17055],...</td>\n",
       "      <td>Crocus sativus [TAX:82528]</td>\n",
       "      <td>Iridaceae (Iris family) Saffron chapiter\\r\\nMa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E00004</td>\n",
       "      <td>Agar (JP17/NF)</td>\n",
       "      <td>Crude drug</td>\n",
       "      <td>Agarose [CPD:C01399]</td>\n",
       "      <td>Gelidium amansii [TAX:2812]</td>\n",
       "      <td>Gelidiaceae Gelidium amansii mucous (freeze dr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E00005</td>\n",
       "      <td>Tragacanth (JP17/NF)</td>\n",
       "      <td>Crude drug</td>\n",
       "      <td>D-Galacturonic acid [CPD:C00333], Tragacanthic...</td>\n",
       "      <td>Astragalus gummifer [TAX:339493], Leguminosae ...</td>\n",
       "      <td>Fabaceae (pea family) Tragacanth stem exudatio...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  E number                   Name                    Category  \\\n",
       "0   E00001  Coptis rhizome (JP17)                  Crude drug   \n",
       "1   E00002    Evodia fruit (JP17)                  Crude drug   \n",
       "2   E00003         Saffron (JP17)  Crude drug; Medicinal herb   \n",
       "3   E00004         Agar (JP17/NF)                  Crude drug   \n",
       "4   E00005   Tragacanth (JP17/NF)                  Crude drug   \n",
       "\n",
       "                                           Component  \\\n",
       "0  Berberine [CPD:C00757], Palmatine [CPD:C05315]...   \n",
       "1  Evodiamine [CPD:C09187], Rutecarpine [CPD:C092...   \n",
       "2  Crocin [CPD:C08589], Picrocrocin [CPD:C17055],...   \n",
       "3                               Agarose [CPD:C01399]   \n",
       "4  D-Galacturonic acid [CPD:C00333], Tragacanthic...   \n",
       "\n",
       "                                              Source  \\\n",
       "0  Coptis japonica [TAX:3442], Coptis chinensis [...   \n",
       "1  Evodia rutaecarpa [TAX:354523], Evodia officin...   \n",
       "2                         Crocus sativus [TAX:82528]   \n",
       "3                        Gelidium amansii [TAX:2812]   \n",
       "4  Astragalus gummifer [TAX:339493], Leguminosae ...   \n",
       "\n",
       "                                             Comment  \n",
       "0  Ranunculaceae (buttercup family) Coptis rhizom...  \n",
       "1  Rutaceae (rue family) Evodia fruit\\r\\nMajor co...  \n",
       "2  Iridaceae (Iris family) Saffron chapiter\\r\\nMa...  \n",
       "3  Gelidiaceae Gelidium amansii mucous (freeze dr...  \n",
       "4  Fabaceae (pea family) Tragacanth stem exudatio...  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_environ_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4\n",
    "def row_to_dict(row):\n",
    "    entry_id = row['E number']\n",
    "    name = row['Name']\n",
    "    category = row['Category']\n",
    "    component = row['Component']\n",
    "    source = row['Source']\n",
    "    comment = row['Comment']\n",
    "    text = f\"name: {name}; category: {category}; component: {comment} comment: {comment}; source: {source}\"\n",
    "\n",
    "    output_dict = {\n",
    "            \"id\": entry_id,\n",
    "            \"input\": text,\n",
    "            \"metadata\": {\n",
    "                \"name\": name, \n",
    "                \"text\": text,\n",
    "                \"component\": component,\n",
    "                \n",
    "            }\n",
    "        }\n",
    "\n",
    "    return output_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "850\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'id': 'E00001',\n",
       " 'input': 'name: Coptis rhizome (JP17); category: Crude drug; component: Ranunculaceae (buttercup family) Coptis rhizome\\r\\nMajor component: Berberine [CPD:C00757] comment: Ranunculaceae (buttercup family) Coptis rhizome\\r\\nMajor component: Berberine [CPD:C00757]; source: Coptis japonica [TAX:3442], Coptis chinensis [TAX:261450], Coptis deltoidea [TAX:261449], Coptis teeta [TAX:261448]',\n",
       " 'metadata': {'name': 'Coptis rhizome (JP17)',\n",
       "  'text': 'name: Coptis rhizome (JP17); category: Crude drug; component: Ranunculaceae (buttercup family) Coptis rhizome\\r\\nMajor component: Berberine [CPD:C00757] comment: Ranunculaceae (buttercup family) Coptis rhizome\\r\\nMajor component: Berberine [CPD:C00757]; source: Coptis japonica [TAX:3442], Coptis chinensis [TAX:261450], Coptis deltoidea [TAX:261449], Coptis teeta [TAX:261448]',\n",
       "  'component': 'Berberine [CPD:C00757], Palmatine [CPD:C05315], Jateorrhizine [CPD:C09553], Coptisine [CPD:C16938], Worenine [CPD:C17083], Magnoflorine [CPD:C09581], Ferulic acid [CPD:C01494], Chlorogenic acid [CPD:C00852], Tetrahydroberberine [CPD:C03329], Tetrahydropalmatine [CPD:C02890]'}}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "environ_preprocess = convert_dataframe_to_list(clean_environ_df)\n",
    "print(len(environ_preprocess))\n",
    "environ_preprocess[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(850, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>input</th>\n",
       "      <th>metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E00001</td>\n",
       "      <td>name: Coptis rhizome (JP17); category: Crude d...</td>\n",
       "      <td>{'name': 'Coptis rhizome (JP17)', 'text': 'nam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E00002</td>\n",
       "      <td>name: Evodia fruit (JP17); category: Crude dru...</td>\n",
       "      <td>{'name': 'Evodia fruit (JP17)', 'text': 'name:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E00003</td>\n",
       "      <td>name: Saffron (JP17); category: Crude drug; Me...</td>\n",
       "      <td>{'name': 'Saffron (JP17)', 'text': 'name: Saff...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E00004</td>\n",
       "      <td>name: Agar (JP17/NF); category: Crude drug; co...</td>\n",
       "      <td>{'name': 'Agar (JP17/NF)', 'text': 'name: Agar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E00005</td>\n",
       "      <td>name: Tragacanth (JP17/NF); category: Crude dr...</td>\n",
       "      <td>{'name': 'Tragacanth (JP17/NF)', 'text': 'name...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                              input  \\\n",
       "0  E00001  name: Coptis rhizome (JP17); category: Crude d...   \n",
       "1  E00002  name: Evodia fruit (JP17); category: Crude dru...   \n",
       "2  E00003  name: Saffron (JP17); category: Crude drug; Me...   \n",
       "3  E00004  name: Agar (JP17/NF); category: Crude drug; co...   \n",
       "4  E00005  name: Tragacanth (JP17/NF); category: Crude dr...   \n",
       "\n",
       "                                            metadata  \n",
       "0  {'name': 'Coptis rhizome (JP17)', 'text': 'nam...  \n",
       "1  {'name': 'Evodia fruit (JP17)', 'text': 'name:...  \n",
       "2  {'name': 'Saffron (JP17)', 'text': 'name: Saff...  \n",
       "3  {'name': 'Agar (JP17/NF)', 'text': 'name: Agar...  \n",
       "4  {'name': 'Tragacanth (JP17/NF)', 'text': 'name...  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "environ_df = pd.DataFrame(environ_preprocess)\n",
    "preprocessed_datasets.append(environ_df)\n",
    "print(environ_df.shape)\n",
    "environ_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kegg Medicus Network Data Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>NAME</th>\n",
       "      <th>DEFINITION</th>\n",
       "      <th>EXPANDED</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENE</th>\n",
       "      <th>VARIANT</th>\n",
       "      <th>METABOLITE</th>\n",
       "      <th>...</th>\n",
       "      <th>ENTRY_link</th>\n",
       "      <th>DEFINITION_link</th>\n",
       "      <th>EXPANDED_link</th>\n",
       "      <th>CLASS_link</th>\n",
       "      <th>DISEASE_link</th>\n",
       "      <th>GENE_link</th>\n",
       "      <th>VARIANT_link</th>\n",
       "      <th>METABOLITE_link</th>\n",
       "      <th>PERTURBANT_link</th>\n",
       "      <th>REFERENCE_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N00001</td>\n",
       "      <td>EGF-EGFR-RAS-ERK signaling pathway</td>\n",
       "      <td>EGF -&gt; EGFR -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; ME...</td>\n",
       "      <td>1950 -&gt; 1956 -&gt; 2885 -&gt; (6654,6655) -&gt; (3265,3...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06260 Colorectal canc...</td>\n",
       "      <td>Reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1950  EGF; epidermal growth factor\\n1956  EGFR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>EGF -&gt; EGFR -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; ME...</td>\n",
       "      <td>&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bget...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/network/nt06210\"&gt;...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bget...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N00002</td>\n",
       "      <td>BCR-ABL fusion kinase to RAS-ERK signaling pat...</td>\n",
       "      <td>BCR-ABL -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt;...</td>\n",
       "      <td>(25v1,25v2) -&gt; 2885 -&gt; (6654,6655) -&gt; (3265,38...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06276 Chronic myeloid...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>H00004  Chronic myeloid leukemia</td>\n",
       "      <td>25  ABL1; ABL proto-oncogene 1, non-receptor t...</td>\n",
       "      <td>25v1 (BCR-ABL)  BCR-ABL1 fusion\\n25v2 (BCR-ABL...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>BCR-ABL -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt;...</td>\n",
       "      <td>(&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/network/nt06210\"&gt;...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/entry/...</td>\n",
       "      <td>&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bget...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N00003</td>\n",
       "      <td>Mutation-activated KIT to RAS-ERK signaling pa...</td>\n",
       "      <td>KIT* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ERK</td>\n",
       "      <td>3815v1 -&gt; 2885 -&gt; (6654,6655) -&gt; (3265,3845,48...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06275 Acute myeloid l...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>H00003  Acute myeloid leukemia</td>\n",
       "      <td>3815  KIT; KIT proto-oncogene receptor tyrosin...</td>\n",
       "      <td>3815v1 (KIT*)  KIT kinase domain mutation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>KIT* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ERK</td>\n",
       "      <td>&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bget...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/network/nt06210\"&gt;...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/entry/...</td>\n",
       "      <td>&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bget...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N00004</td>\n",
       "      <td>Duplication or mutation-activated FLT3 to RAS-...</td>\n",
       "      <td>FLT3* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ERK</td>\n",
       "      <td>(2322v2,2322v1) -&gt; 2885 -&gt; (6654,6655) -&gt; (326...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06275 Acute myeloid l...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>H00003  Acute myeloid leukemia</td>\n",
       "      <td>2322  FLT3; fms related tyrosine kinase 3\\n288...</td>\n",
       "      <td>2322v2 (FLT3*)  FLT3 internal tandem duplicati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>FLT3* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ERK</td>\n",
       "      <td>(&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/network/nt06210\"&gt;...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/entry/...</td>\n",
       "      <td>&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bget...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N00005</td>\n",
       "      <td>Mutation-activated MET to RAS-ERK signaling pa...</td>\n",
       "      <td>MET* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ER...</td>\n",
       "      <td>4233v1 -&gt; 2885 -&gt; (6654,6655) -&gt; (3265,3845,48...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06263 Hepatocellular ...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>H00048  Hepatocellular carcinoma\\nH00021  Rena...</td>\n",
       "      <td>4233  MET; MET proto-oncogene, receptor tyrosi...</td>\n",
       "      <td>4233v1 (MET*)  MET kinase domain mutation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/dbget-bin/www_bge...</td>\n",
       "      <td>MET* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ER...</td>\n",
       "      <td>&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bget...</td>\n",
       "      <td>&lt;a href=\"https://www.kegg.jp/network/nt06210\"&gt;...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/entry/...</td>\n",
       "      <td>&lt;a href=\"http://www.kegg.jp/dbget-bin/www_bget...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ENTRY                                               NAME  \\\n",
       "0  N00001                 EGF-EGFR-RAS-ERK signaling pathway   \n",
       "1  N00002  BCR-ABL fusion kinase to RAS-ERK signaling pat...   \n",
       "2  N00003  Mutation-activated KIT to RAS-ERK signaling pa...   \n",
       "3  N00004  Duplication or mutation-activated FLT3 to RAS-...   \n",
       "4  N00005  Mutation-activated MET to RAS-ERK signaling pa...   \n",
       "\n",
       "                                          DEFINITION  \\\n",
       "0  EGF -> EGFR -> GRB2 -> SOS -> RAS -> RAF -> ME...   \n",
       "1  BCR-ABL -> GRB2 -> SOS -> RAS -> RAF -> MEK ->...   \n",
       "2    KIT* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ERK   \n",
       "3   FLT3* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ERK   \n",
       "4  MET* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ER...   \n",
       "\n",
       "                                            EXPANDED  \\\n",
       "0  1950 -> 1956 -> 2885 -> (6654,6655) -> (3265,3...   \n",
       "1  (25v1,25v2) -> 2885 -> (6654,6655) -> (3265,38...   \n",
       "2  3815v1 -> 2885 -> (6654,6655) -> (3265,3845,48...   \n",
       "3  (2322v2,2322v1) -> 2885 -> (6654,6655) -> (326...   \n",
       "4  4233v1 -> 2885 -> (6654,6655) -> (3265,3845,48...   \n",
       "\n",
       "                                               CLASS       TYPE  \\\n",
       "0  nt06210 ERK signaling\\nnt06260 Colorectal canc...  Reference   \n",
       "1  nt06210 ERK signaling\\nnt06276 Chronic myeloid...    Variant   \n",
       "2  nt06210 ERK signaling\\nnt06275 Acute myeloid l...    Variant   \n",
       "3  nt06210 ERK signaling\\nnt06275 Acute myeloid l...    Variant   \n",
       "4  nt06210 ERK signaling\\nnt06263 Hepatocellular ...    Variant   \n",
       "\n",
       "                                             DISEASE  \\\n",
       "0                                                NaN   \n",
       "1                   H00004  Chronic myeloid leukemia   \n",
       "2                     H00003  Acute myeloid leukemia   \n",
       "3                     H00003  Acute myeloid leukemia   \n",
       "4  H00048  Hepatocellular carcinoma\\nH00021  Rena...   \n",
       "\n",
       "                                                GENE  \\\n",
       "0  1950  EGF; epidermal growth factor\\n1956  EGFR...   \n",
       "1  25  ABL1; ABL proto-oncogene 1, non-receptor t...   \n",
       "2  3815  KIT; KIT proto-oncogene receptor tyrosin...   \n",
       "3  2322  FLT3; fms related tyrosine kinase 3\\n288...   \n",
       "4  4233  MET; MET proto-oncogene, receptor tyrosi...   \n",
       "\n",
       "                                             VARIANT METABOLITE  ...  \\\n",
       "0                                                NaN        NaN  ...   \n",
       "1  25v1 (BCR-ABL)  BCR-ABL1 fusion\\n25v2 (BCR-ABL...        NaN  ...   \n",
       "2          3815v1 (KIT*)  KIT kinase domain mutation        NaN  ...   \n",
       "3  2322v2 (FLT3*)  FLT3 internal tandem duplicati...        NaN  ...   \n",
       "4          4233v1 (MET*)  MET kinase domain mutation        NaN  ...   \n",
       "\n",
       "                                          ENTRY_link  \\\n",
       "0  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "1  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "2  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "3  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "4  <a href=\"https://www.kegg.jp/dbget-bin/www_bge...   \n",
       "\n",
       "                                     DEFINITION_link  \\\n",
       "0  EGF -> EGFR -> GRB2 -> SOS -> RAS -> RAF -> ME...   \n",
       "1  BCR-ABL -> GRB2 -> SOS -> RAS -> RAF -> MEK ->...   \n",
       "2    KIT* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ERK   \n",
       "3   FLT3* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ERK   \n",
       "4  MET* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ER...   \n",
       "\n",
       "                                       EXPANDED_link  \\\n",
       "0  <a href=\"http://www.kegg.jp/dbget-bin/www_bget...   \n",
       "1  (<a href=\"http://www.kegg.jp/dbget-bin/www_bge...   \n",
       "2  <a href=\"http://www.kegg.jp/dbget-bin/www_bget...   \n",
       "3  (<a href=\"http://www.kegg.jp/dbget-bin/www_bge...   \n",
       "4  <a href=\"http://www.kegg.jp/dbget-bin/www_bget...   \n",
       "\n",
       "                                          CLASS_link  \\\n",
       "0  <a href=\"https://www.kegg.jp/network/nt06210\">...   \n",
       "1  <a href=\"https://www.kegg.jp/network/nt06210\">...   \n",
       "2  <a href=\"https://www.kegg.jp/network/nt06210\">...   \n",
       "3  <a href=\"https://www.kegg.jp/network/nt06210\">...   \n",
       "4  <a href=\"https://www.kegg.jp/network/nt06210\">...   \n",
       "\n",
       "                                        DISEASE_link  \\\n",
       "0                                                NaN   \n",
       "1  <a href=\"http://togodb.biosciencedbc.jp/entry/...   \n",
       "2  <a href=\"http://togodb.biosciencedbc.jp/entry/...   \n",
       "3  <a href=\"http://togodb.biosciencedbc.jp/entry/...   \n",
       "4  <a href=\"http://togodb.biosciencedbc.jp/entry/...   \n",
       "\n",
       "                                           GENE_link  \\\n",
       "0  <a href=\"http://www.kegg.jp/dbget-bin/www_bget...   \n",
       "1  <a href=\"http://www.kegg.jp/dbget-bin/www_bget...   \n",
       "2  <a href=\"http://www.kegg.jp/dbget-bin/www_bget...   \n",
       "3  <a href=\"http://www.kegg.jp/dbget-bin/www_bget...   \n",
       "4  <a href=\"http://www.kegg.jp/dbget-bin/www_bget...   \n",
       "\n",
       "                                        VARIANT_link METABOLITE_link  \\\n",
       "0                                                NaN             NaN   \n",
       "1  <a href=\"http://togodb.biosciencedbc.jp/togodb...             NaN   \n",
       "2  <a href=\"http://togodb.biosciencedbc.jp/togodb...             NaN   \n",
       "3  <a href=\"http://togodb.biosciencedbc.jp/togodb...             NaN   \n",
       "4  <a href=\"http://togodb.biosciencedbc.jp/togodb...             NaN   \n",
       "\n",
       "  PERTURBANT_link                                     REFERENCE_link  \n",
       "0             NaN  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "1             NaN  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "2             NaN  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "3             NaN  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "4             NaN  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_network_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1241, 22)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_network_df.shape "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude = ['ENTRY_link','VARIANT','METABOLITE','PERTURBANT','DEFINITION_link', 'EXPANDED_link', 'CLASS_link', 'DISEASE_link', 'GENE_link', 'VARIANT_link',\n",
    "           'METABOLITE_link', 'PERTURBANT_link','REFERENCE_link']\n",
    "clean_network_df = raw_network_df[[col for col in raw_network_df.columns if col not in exclude]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>NAME</th>\n",
       "      <th>DEFINITION</th>\n",
       "      <th>EXPANDED</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENE</th>\n",
       "      <th>REFERENCE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N00001</td>\n",
       "      <td>EGF-EGFR-RAS-ERK signaling pathway</td>\n",
       "      <td>EGF -&gt; EGFR -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; ME...</td>\n",
       "      <td>1950 -&gt; 1956 -&gt; 2885 -&gt; (6654,6655) -&gt; (3265,3...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06260 Colorectal canc...</td>\n",
       "      <td>Reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1950  EGF; epidermal growth factor\\n1956  EGFR...</td>\n",
       "      <td>PMID:17409820\\nAUTHORS   Molina JR, Adjei AA\\n...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N00002</td>\n",
       "      <td>BCR-ABL fusion kinase to RAS-ERK signaling pat...</td>\n",
       "      <td>BCR-ABL -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt;...</td>\n",
       "      <td>(25v1,25v2) -&gt; 2885 -&gt; (6654,6655) -&gt; (3265,38...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06276 Chronic myeloid...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>H00004  Chronic myeloid leukemia</td>\n",
       "      <td>25  ABL1; ABL proto-oncogene 1, non-receptor t...</td>\n",
       "      <td>PMID:10403855\\nAUTHORS   Faderl S, Talpaz M, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N00003</td>\n",
       "      <td>Mutation-activated KIT to RAS-ERK signaling pa...</td>\n",
       "      <td>KIT* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ERK</td>\n",
       "      <td>3815v1 -&gt; 2885 -&gt; (6654,6655) -&gt; (3265,3845,48...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06275 Acute myeloid l...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>H00003  Acute myeloid leukemia</td>\n",
       "      <td>3815  KIT; KIT proto-oncogene receptor tyrosin...</td>\n",
       "      <td>PMID:23678293\\nAUTHORS   Liang J, Wu YL, Chen ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N00004</td>\n",
       "      <td>Duplication or mutation-activated FLT3 to RAS-...</td>\n",
       "      <td>FLT3* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ERK</td>\n",
       "      <td>(2322v2,2322v1) -&gt; 2885 -&gt; (6654,6655) -&gt; (326...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06275 Acute myeloid l...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>H00003  Acute myeloid leukemia</td>\n",
       "      <td>2322  FLT3; fms related tyrosine kinase 3\\n288...</td>\n",
       "      <td>PMID:12951584\\nAUTHORS   Stirewalt DL, Radich ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N00005</td>\n",
       "      <td>Mutation-activated MET to RAS-ERK signaling pa...</td>\n",
       "      <td>MET* -&gt; GRB2 -&gt; SOS -&gt; RAS -&gt; RAF -&gt; MEK -&gt; ER...</td>\n",
       "      <td>4233v1 -&gt; 2885 -&gt; (6654,6655) -&gt; (3265,3845,48...</td>\n",
       "      <td>nt06210 ERK signaling\\nnt06263 Hepatocellular ...</td>\n",
       "      <td>Variant</td>\n",
       "      <td>H00048  Hepatocellular carcinoma\\nH00021  Rena...</td>\n",
       "      <td>4233  MET; MET proto-oncogene, receptor tyrosi...</td>\n",
       "      <td>PMID:14685170\\nAUTHORS   Birchmeier C, Birchme...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ENTRY                                               NAME  \\\n",
       "0  N00001                 EGF-EGFR-RAS-ERK signaling pathway   \n",
       "1  N00002  BCR-ABL fusion kinase to RAS-ERK signaling pat...   \n",
       "2  N00003  Mutation-activated KIT to RAS-ERK signaling pa...   \n",
       "3  N00004  Duplication or mutation-activated FLT3 to RAS-...   \n",
       "4  N00005  Mutation-activated MET to RAS-ERK signaling pa...   \n",
       "\n",
       "                                          DEFINITION  \\\n",
       "0  EGF -> EGFR -> GRB2 -> SOS -> RAS -> RAF -> ME...   \n",
       "1  BCR-ABL -> GRB2 -> SOS -> RAS -> RAF -> MEK ->...   \n",
       "2    KIT* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ERK   \n",
       "3   FLT3* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ERK   \n",
       "4  MET* -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ER...   \n",
       "\n",
       "                                            EXPANDED  \\\n",
       "0  1950 -> 1956 -> 2885 -> (6654,6655) -> (3265,3...   \n",
       "1  (25v1,25v2) -> 2885 -> (6654,6655) -> (3265,38...   \n",
       "2  3815v1 -> 2885 -> (6654,6655) -> (3265,3845,48...   \n",
       "3  (2322v2,2322v1) -> 2885 -> (6654,6655) -> (326...   \n",
       "4  4233v1 -> 2885 -> (6654,6655) -> (3265,3845,48...   \n",
       "\n",
       "                                               CLASS       TYPE  \\\n",
       "0  nt06210 ERK signaling\\nnt06260 Colorectal canc...  Reference   \n",
       "1  nt06210 ERK signaling\\nnt06276 Chronic myeloid...    Variant   \n",
       "2  nt06210 ERK signaling\\nnt06275 Acute myeloid l...    Variant   \n",
       "3  nt06210 ERK signaling\\nnt06275 Acute myeloid l...    Variant   \n",
       "4  nt06210 ERK signaling\\nnt06263 Hepatocellular ...    Variant   \n",
       "\n",
       "                                             DISEASE  \\\n",
       "0                                                NaN   \n",
       "1                   H00004  Chronic myeloid leukemia   \n",
       "2                     H00003  Acute myeloid leukemia   \n",
       "3                     H00003  Acute myeloid leukemia   \n",
       "4  H00048  Hepatocellular carcinoma\\nH00021  Rena...   \n",
       "\n",
       "                                                GENE  \\\n",
       "0  1950  EGF; epidermal growth factor\\n1956  EGFR...   \n",
       "1  25  ABL1; ABL proto-oncogene 1, non-receptor t...   \n",
       "2  3815  KIT; KIT proto-oncogene receptor tyrosin...   \n",
       "3  2322  FLT3; fms related tyrosine kinase 3\\n288...   \n",
       "4  4233  MET; MET proto-oncogene, receptor tyrosi...   \n",
       "\n",
       "                                           REFERENCE  \n",
       "0  PMID:17409820\\nAUTHORS   Molina JR, Adjei AA\\n...  \n",
       "1  PMID:10403855\\nAUTHORS   Faderl S, Talpaz M, E...  \n",
       "2  PMID:23678293\\nAUTHORS   Liang J, Wu YL, Chen ...  \n",
       "3  PMID:12951584\\nAUTHORS   Stirewalt DL, Radich ...  \n",
       "4  PMID:14685170\\nAUTHORS   Birchmeier C, Birchme...  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_network_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENTRY                0\n",
       "NAME                 0\n",
       "DEFINITION           0\n",
       "EXPANDED             0\n",
       "CLASS               24\n",
       "TYPE                 0\n",
       "DISEASE            361\n",
       "GENE                 4\n",
       "VARIANT            867\n",
       "METABOLITE         924\n",
       "PERTURBANT         662\n",
       "REFERENCE           76\n",
       "ENTRY_link           0\n",
       "DEFINITION_link      0\n",
       "EXPANDED_link        0\n",
       "CLASS_link          24\n",
       "DISEASE_link       361\n",
       "GENE_link            4\n",
       "VARIANT_link       867\n",
       "METABOLITE_link    924\n",
       "PERTURBANT_link    662\n",
       "REFERENCE_link      76\n",
       "dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_network_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENTRY', 'NAME', 'DEFINITION', 'EXPANDED', 'CLASS', 'TYPE', 'DISEASE',\n",
       "       'GENE', 'REFERENCE'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_network_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "#5\n",
    "def row_to_dict(row):\n",
    "    entry_id = row['ENTRY']\n",
    "    name = row['NAME']\n",
    "    definition = row['DEFINITION']\n",
    "    expanded = row['EXPANDED']\n",
    "    item_class = row['CLASS']\n",
    "    item_type = row['TYPE']\n",
    "    disease = row['DISEASE']\n",
    "    gene = row['GENE']\n",
    "    text = f\"name: {name}; definition: {definition}; expanded: {expanded}; class: {item_class}, type: {item_type}\"\n",
    "\n",
    "    output_dict = {\n",
    "            \"id\": entry_id,\n",
    "            \"input\": text,\n",
    "            \"metadata\": {\n",
    "                \"name\": name,\n",
    "                \"text\": text,\n",
    "                \"name\": name,\n",
    "                \"class\": item_class,\n",
    "                \"comment\": item_type,\n",
    "                \"disease\": disease,\n",
    "                \"gene\":gene\n",
    "            }\n",
    "        }\n",
    "\n",
    "    return output_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1241\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'id': 'N00001',\n",
       " 'input': 'name: EGF-EGFR-RAS-ERK signaling pathway; definition: EGF -> EGFR -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ERK -> CCND1; expanded: 1950 -> 1956 -> 2885 -> (6654,6655) -> (3265,3845,4893) -> (369,673,5894) -> (5604,5605) -> (5594,5595) -> 595; class: nt06210 ERK signaling\\nnt06260 Colorectal cancer\\nnt06263 Hepatocellular carcinoma\\nnt06265 Bladder cancer\\nnt06266 Non-small cell lung cancer\\nnt06268 Melanoma\\nnt06270 Breast cancer\\nnt06271 Endometrial cancer\\nnt06273 Glioma\\nnt06274 Thyroid cancer\\nnt06276 Chronic myeloid leukemia\\nnt06110 MAPK signaling (viruses and bacteria)\\nnt06162 Hepatitis B virus (HBV)\\nnt06163 Hepatitis C virus (HCV)\\nnt06170 Influenza A virus (IAV)\\nnt06167 Human cytomegalovirus (HCMV)\\nnt06166 Human papillomavirus (HPV), type: Reference',\n",
       " 'metadata': {'name': 'EGF-EGFR-RAS-ERK signaling pathway',\n",
       "  'text': 'name: EGF-EGFR-RAS-ERK signaling pathway; definition: EGF -> EGFR -> GRB2 -> SOS -> RAS -> RAF -> MEK -> ERK -> CCND1; expanded: 1950 -> 1956 -> 2885 -> (6654,6655) -> (3265,3845,4893) -> (369,673,5894) -> (5604,5605) -> (5594,5595) -> 595; class: nt06210 ERK signaling\\nnt06260 Colorectal cancer\\nnt06263 Hepatocellular carcinoma\\nnt06265 Bladder cancer\\nnt06266 Non-small cell lung cancer\\nnt06268 Melanoma\\nnt06270 Breast cancer\\nnt06271 Endometrial cancer\\nnt06273 Glioma\\nnt06274 Thyroid cancer\\nnt06276 Chronic myeloid leukemia\\nnt06110 MAPK signaling (viruses and bacteria)\\nnt06162 Hepatitis B virus (HBV)\\nnt06163 Hepatitis C virus (HCV)\\nnt06170 Influenza A virus (IAV)\\nnt06167 Human cytomegalovirus (HCMV)\\nnt06166 Human papillomavirus (HPV), type: Reference',\n",
       "  'class': 'nt06210 ERK signaling\\nnt06260 Colorectal cancer\\nnt06263 Hepatocellular carcinoma\\nnt06265 Bladder cancer\\nnt06266 Non-small cell lung cancer\\nnt06268 Melanoma\\nnt06270 Breast cancer\\nnt06271 Endometrial cancer\\nnt06273 Glioma\\nnt06274 Thyroid cancer\\nnt06276 Chronic myeloid leukemia\\nnt06110 MAPK signaling (viruses and bacteria)\\nnt06162 Hepatitis B virus (HBV)\\nnt06163 Hepatitis C virus (HCV)\\nnt06170 Influenza A virus (IAV)\\nnt06167 Human cytomegalovirus (HCMV)\\nnt06166 Human papillomavirus (HPV)',\n",
       "  'comment': 'Reference',\n",
       "  'disease': nan,\n",
       "  'gene': '1950  EGF; epidermal growth factor\\n1956  EGFR; epidermal growth factor receptor\\n2885  GRB2; growth factor receptor bound protein 2\\n6654  SOS1; SOS Ras/Rac guanine nucleotide exchange factor 1\\n6655  SOS2; SOS Ras/Rho guanine nucleotide exchange factor 2\\n3265  HRAS; HRas proto-oncogene, GTPase\\n3845  KRAS; KRAS proto-oncogene, GTPase\\n4893  NRAS; NRAS proto-oncogene, GTPase\\n369  ARAF; A-Raf proto-oncogene, serine/threonine kinase\\n673  BRAF; B-Raf proto-oncogene, serine/threonine kinase\\n5894  RAF1; Raf-1 proto-oncogene, serine/threonine kinase\\n5604  MAP2K1; mitogen-activated protein kinase kinase 1\\n5605  MAP2K2; mitogen-activated protein kinase kinase 2\\n5594  MAPK1; mitogen-activated protein kinase 1\\n5595  MAPK3; mitogen-activated protein kinase 3\\n595  CCND1; cyclin D1'}}"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "network_preprocess = convert_dataframe_to_list(clean_network_df)\n",
    "print(len(network_preprocess))\n",
    "network_preprocess[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1241, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>input</th>\n",
       "      <th>metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N00001</td>\n",
       "      <td>name: EGF-EGFR-RAS-ERK signaling pathway; defi...</td>\n",
       "      <td>{'name': 'EGF-EGFR-RAS-ERK signaling pathway',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N00002</td>\n",
       "      <td>name: BCR-ABL fusion kinase to RAS-ERK signali...</td>\n",
       "      <td>{'name': 'BCR-ABL fusion kinase to RAS-ERK sig...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N00003</td>\n",
       "      <td>name: Mutation-activated KIT to RAS-ERK signal...</td>\n",
       "      <td>{'name': 'Mutation-activated KIT to RAS-ERK si...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N00004</td>\n",
       "      <td>name: Duplication or mutation-activated FLT3 t...</td>\n",
       "      <td>{'name': 'Duplication or mutation-activated FL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N00005</td>\n",
       "      <td>name: Mutation-activated MET to RAS-ERK signal...</td>\n",
       "      <td>{'name': 'Mutation-activated MET to RAS-ERK si...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                              input  \\\n",
       "0  N00001  name: EGF-EGFR-RAS-ERK signaling pathway; defi...   \n",
       "1  N00002  name: BCR-ABL fusion kinase to RAS-ERK signali...   \n",
       "2  N00003  name: Mutation-activated KIT to RAS-ERK signal...   \n",
       "3  N00004  name: Duplication or mutation-activated FLT3 t...   \n",
       "4  N00005  name: Mutation-activated MET to RAS-ERK signal...   \n",
       "\n",
       "                                            metadata  \n",
       "0  {'name': 'EGF-EGFR-RAS-ERK signaling pathway',...  \n",
       "1  {'name': 'BCR-ABL fusion kinase to RAS-ERK sig...  \n",
       "2  {'name': 'Mutation-activated KIT to RAS-ERK si...  \n",
       "3  {'name': 'Duplication or mutation-activated FL...  \n",
       "4  {'name': 'Mutation-activated MET to RAS-ERK si...  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "network_df = pd.DataFrame(network_preprocess)\n",
    "preprocessed_datasets.append(network_df)\n",
    "print(network_df.shape)\n",
    "network_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Kegg Medicus Variant Data Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>NAME</th>\n",
       "      <th>GENE</th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>VARIATION</th>\n",
       "      <th>NETWORK</th>\n",
       "      <th>ELEMENT</th>\n",
       "      <th>REFERENCE</th>\n",
       "      <th>ENTRY_link</th>\n",
       "      <th>GENE_link</th>\n",
       "      <th>VARIATION_link</th>\n",
       "      <th>NETWORK_link</th>\n",
       "      <th>ELEMENT_link</th>\n",
       "      <th>REFERENCE_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10026v1</td>\n",
       "      <td>PIGK deficiency</td>\n",
       "      <td>PIGK  phosphatidylinositol glycan anchor biosy...</td>\n",
       "      <td>hsa_var Human gene variants (Homo sapiens)</td>\n",
       "      <td>mutations\\nOmimVar: 605087</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:32220290\\nAUTHORS   Nguyen TTM, Murakami ...</td>\n",
       "      <td>&lt;a href=\"https://www.genome.jp/dbget-bin/www_b...</td>\n",
       "      <td>PIGK  phosphatidylinositol glycan anchor biosy...</td>\n",
       "      <td>mutations\\nOmimVar: &lt;a href=\"https://omim.org/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100v1</td>\n",
       "      <td>ADA deficiency</td>\n",
       "      <td>ADA  adenosine deaminase [KO:K01488]</td>\n",
       "      <td>hsa_var Human gene variants (Homo sapiens)</td>\n",
       "      <td>mutations\\nOmimVar: 608958\\nDRUG_TARGET Elapeg...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;a href=\"https://www.genome.jp/dbget-bin/www_b...</td>\n",
       "      <td>ADA  adenosine deaminase [KO:&lt;a href=\"http://w...</td>\n",
       "      <td>mutations\\nOmimVar: &lt;a href=\"https://omim.org/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10133v1</td>\n",
       "      <td>OPTN mutation</td>\n",
       "      <td>OPTN  optineurin [KO:K19946]</td>\n",
       "      <td>hsa_var Human gene variants (Homo sapiens)</td>\n",
       "      <td>mutations\\nOmimVar: 10133</td>\n",
       "      <td>nt06421  Mitophagy\\nnt06464  Amyotrophic later...</td>\n",
       "      <td>N01138  Mutation-inactivated OPTN to PINK-Park...</td>\n",
       "      <td>PMID:20428114\\nAUTHORS   Maruyama H, Morino H,...</td>\n",
       "      <td>&lt;a href=\"https://www.genome.jp/dbget-bin/www_b...</td>\n",
       "      <td>OPTN  optineurin [KO:&lt;a href=\"http://www.kegg....</td>\n",
       "      <td>mutations\\nOmimVar: &lt;a href=\"https://omim.org/...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10195v1</td>\n",
       "      <td>ALG3 deficiency</td>\n",
       "      <td>ALG3  ALG3 alpha-1,3- mannosyltransferase [KO:...</td>\n",
       "      <td>hsa_var Human gene variants (Homo sapiens)</td>\n",
       "      <td>mutations\\nOmimVar: 608750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:15108280\\nAUTHORS   Denecke J, Kranz C, K...</td>\n",
       "      <td>&lt;a href=\"https://www.genome.jp/dbget-bin/www_b...</td>\n",
       "      <td>ALG3  ALG3 alpha-1,3- mannosyltransferase [KO:...</td>\n",
       "      <td>mutations\\nOmimVar: &lt;a href=\"https://omim.org/...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1019v1</td>\n",
       "      <td>CDK4 amplification</td>\n",
       "      <td>CDK4  cyclin dependent kinase 4 [KO:K02089]</td>\n",
       "      <td>hsa_var Human gene variants (Homo sapiens)</td>\n",
       "      <td>amplification\\nClinVar: 268075 150740\\ndbVar: ...</td>\n",
       "      <td>nt06230  Cell cycle\\nnt06273  Glioma</td>\n",
       "      <td>N00072  Amplified CDK4 to cell cycle G1/S</td>\n",
       "      <td>PMID:7987821\\nAUTHORS   Schmidt EE, Ichimura K...</td>\n",
       "      <td>&lt;a href=\"https://www.genome.jp/dbget-bin/www_b...</td>\n",
       "      <td>CDK4  cyclin dependent kinase 4 [KO:&lt;a href=\"h...</td>\n",
       "      <td>amplification\\nClinVar: &lt;a href=\"https://www.n...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>&lt;a href=\"http://togodb.biosciencedbc.jp/togodb...</td>\n",
       "      <td>PMID:&lt;a href=\"https://pubmed.ncbi.nlm.nih.gov/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ENTRY                NAME  \\\n",
       "0  10026v1     PIGK deficiency   \n",
       "1    100v1      ADA deficiency   \n",
       "2  10133v1       OPTN mutation   \n",
       "3  10195v1     ALG3 deficiency   \n",
       "4   1019v1  CDK4 amplification   \n",
       "\n",
       "                                                GENE  \\\n",
       "0  PIGK  phosphatidylinositol glycan anchor biosy...   \n",
       "1               ADA  adenosine deaminase [KO:K01488]   \n",
       "2                       OPTN  optineurin [KO:K19946]   \n",
       "3  ALG3  ALG3 alpha-1,3- mannosyltransferase [KO:...   \n",
       "4        CDK4  cyclin dependent kinase 4 [KO:K02089]   \n",
       "\n",
       "                                     ORGANISM  \\\n",
       "0  hsa_var Human gene variants (Homo sapiens)   \n",
       "1  hsa_var Human gene variants (Homo sapiens)   \n",
       "2  hsa_var Human gene variants (Homo sapiens)   \n",
       "3  hsa_var Human gene variants (Homo sapiens)   \n",
       "4  hsa_var Human gene variants (Homo sapiens)   \n",
       "\n",
       "                                           VARIATION  \\\n",
       "0                         mutations\\nOmimVar: 605087   \n",
       "1  mutations\\nOmimVar: 608958\\nDRUG_TARGET Elapeg...   \n",
       "2                          mutations\\nOmimVar: 10133   \n",
       "3                         mutations\\nOmimVar: 608750   \n",
       "4  amplification\\nClinVar: 268075 150740\\ndbVar: ...   \n",
       "\n",
       "                                             NETWORK  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  nt06421  Mitophagy\\nnt06464  Amyotrophic later...   \n",
       "3                                                NaN   \n",
       "4               nt06230  Cell cycle\\nnt06273  Glioma   \n",
       "\n",
       "                                             ELEMENT  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  N01138  Mutation-inactivated OPTN to PINK-Park...   \n",
       "3                                                NaN   \n",
       "4          N00072  Amplified CDK4 to cell cycle G1/S   \n",
       "\n",
       "                                           REFERENCE  \\\n",
       "0  PMID:32220290\\nAUTHORS   Nguyen TTM, Murakami ...   \n",
       "1                                                NaN   \n",
       "2  PMID:20428114\\nAUTHORS   Maruyama H, Morino H,...   \n",
       "3  PMID:15108280\\nAUTHORS   Denecke J, Kranz C, K...   \n",
       "4  PMID:7987821\\nAUTHORS   Schmidt EE, Ichimura K...   \n",
       "\n",
       "                                          ENTRY_link  \\\n",
       "0  <a href=\"https://www.genome.jp/dbget-bin/www_b...   \n",
       "1  <a href=\"https://www.genome.jp/dbget-bin/www_b...   \n",
       "2  <a href=\"https://www.genome.jp/dbget-bin/www_b...   \n",
       "3  <a href=\"https://www.genome.jp/dbget-bin/www_b...   \n",
       "4  <a href=\"https://www.genome.jp/dbget-bin/www_b...   \n",
       "\n",
       "                                           GENE_link  \\\n",
       "0  PIGK  phosphatidylinositol glycan anchor biosy...   \n",
       "1  ADA  adenosine deaminase [KO:<a href=\"http://w...   \n",
       "2  OPTN  optineurin [KO:<a href=\"http://www.kegg....   \n",
       "3  ALG3  ALG3 alpha-1,3- mannosyltransferase [KO:...   \n",
       "4  CDK4  cyclin dependent kinase 4 [KO:<a href=\"h...   \n",
       "\n",
       "                                      VARIATION_link  \\\n",
       "0  mutations\\nOmimVar: <a href=\"https://omim.org/...   \n",
       "1  mutations\\nOmimVar: <a href=\"https://omim.org/...   \n",
       "2  mutations\\nOmimVar: <a href=\"https://omim.org/...   \n",
       "3  mutations\\nOmimVar: <a href=\"https://omim.org/...   \n",
       "4  amplification\\nClinVar: <a href=\"https://www.n...   \n",
       "\n",
       "                                        NETWORK_link  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "3                                                NaN   \n",
       "4  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "\n",
       "                                        ELEMENT_link  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "3                                                NaN   \n",
       "4  <a href=\"http://togodb.biosciencedbc.jp/togodb...   \n",
       "\n",
       "                                      REFERENCE_link  \n",
       "0  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "1                                                NaN  \n",
       "2  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "3  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  \n",
       "4  PMID:<a href=\"https://pubmed.ncbi.nlm.nih.gov/...  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_variant_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(458, 14)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_variant_df.shape "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENTRY               0\n",
       "NAME                0\n",
       "GENE                0\n",
       "ORGANISM            0\n",
       "VARIATION           0\n",
       "NETWORK           210\n",
       "ELEMENT           198\n",
       "REFERENCE          20\n",
       "ENTRY_link          0\n",
       "GENE_link           0\n",
       "VARIATION_link      0\n",
       "NETWORK_link      210\n",
       "ELEMENT_link      198\n",
       "REFERENCE_link     20\n",
       "dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_variant_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude = ['ENTRY_link', 'NETWORK_link', 'GENE_link','ORGANISM', 'VARIATION_link',\n",
    "           'ELEMENT_link', 'REFERENCE_link']\n",
    "clean_variant_df = raw_variant_df[[col for col in raw_variant_df.columns if col not in exclude]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>NAME</th>\n",
       "      <th>GENE</th>\n",
       "      <th>VARIATION</th>\n",
       "      <th>NETWORK</th>\n",
       "      <th>ELEMENT</th>\n",
       "      <th>REFERENCE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10026v1</td>\n",
       "      <td>PIGK deficiency</td>\n",
       "      <td>PIGK  phosphatidylinositol glycan anchor biosy...</td>\n",
       "      <td>mutations\\nOmimVar: 605087</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:32220290\\nAUTHORS   Nguyen TTM, Murakami ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100v1</td>\n",
       "      <td>ADA deficiency</td>\n",
       "      <td>ADA  adenosine deaminase [KO:K01488]</td>\n",
       "      <td>mutations\\nOmimVar: 608958\\nDRUG_TARGET Elapeg...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10133v1</td>\n",
       "      <td>OPTN mutation</td>\n",
       "      <td>OPTN  optineurin [KO:K19946]</td>\n",
       "      <td>mutations\\nOmimVar: 10133</td>\n",
       "      <td>nt06421  Mitophagy\\nnt06464  Amyotrophic later...</td>\n",
       "      <td>N01138  Mutation-inactivated OPTN to PINK-Park...</td>\n",
       "      <td>PMID:20428114\\nAUTHORS   Maruyama H, Morino H,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10195v1</td>\n",
       "      <td>ALG3 deficiency</td>\n",
       "      <td>ALG3  ALG3 alpha-1,3- mannosyltransferase [KO:...</td>\n",
       "      <td>mutations\\nOmimVar: 608750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMID:15108280\\nAUTHORS   Denecke J, Kranz C, K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1019v1</td>\n",
       "      <td>CDK4 amplification</td>\n",
       "      <td>CDK4  cyclin dependent kinase 4 [KO:K02089]</td>\n",
       "      <td>amplification\\nClinVar: 268075 150740\\ndbVar: ...</td>\n",
       "      <td>nt06230  Cell cycle\\nnt06273  Glioma</td>\n",
       "      <td>N00072  Amplified CDK4 to cell cycle G1/S</td>\n",
       "      <td>PMID:7987821\\nAUTHORS   Schmidt EE, Ichimura K...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ENTRY                NAME  \\\n",
       "0  10026v1     PIGK deficiency   \n",
       "1    100v1      ADA deficiency   \n",
       "2  10133v1       OPTN mutation   \n",
       "3  10195v1     ALG3 deficiency   \n",
       "4   1019v1  CDK4 amplification   \n",
       "\n",
       "                                                GENE  \\\n",
       "0  PIGK  phosphatidylinositol glycan anchor biosy...   \n",
       "1               ADA  adenosine deaminase [KO:K01488]   \n",
       "2                       OPTN  optineurin [KO:K19946]   \n",
       "3  ALG3  ALG3 alpha-1,3- mannosyltransferase [KO:...   \n",
       "4        CDK4  cyclin dependent kinase 4 [KO:K02089]   \n",
       "\n",
       "                                           VARIATION  \\\n",
       "0                         mutations\\nOmimVar: 605087   \n",
       "1  mutations\\nOmimVar: 608958\\nDRUG_TARGET Elapeg...   \n",
       "2                          mutations\\nOmimVar: 10133   \n",
       "3                         mutations\\nOmimVar: 608750   \n",
       "4  amplification\\nClinVar: 268075 150740\\ndbVar: ...   \n",
       "\n",
       "                                             NETWORK  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  nt06421  Mitophagy\\nnt06464  Amyotrophic later...   \n",
       "3                                                NaN   \n",
       "4               nt06230  Cell cycle\\nnt06273  Glioma   \n",
       "\n",
       "                                             ELEMENT  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  N01138  Mutation-inactivated OPTN to PINK-Park...   \n",
       "3                                                NaN   \n",
       "4          N00072  Amplified CDK4 to cell cycle G1/S   \n",
       "\n",
       "                                           REFERENCE  \n",
       "0  PMID:32220290\\nAUTHORS   Nguyen TTM, Murakami ...  \n",
       "1                                                NaN  \n",
       "2  PMID:20428114\\nAUTHORS   Maruyama H, Morino H,...  \n",
       "3  PMID:15108280\\nAUTHORS   Denecke J, Kranz C, K...  \n",
       "4  PMID:7987821\\nAUTHORS   Schmidt EE, Ichimura K...  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_variant_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENTRY', 'NAME', 'GENE', 'VARIATION', 'NETWORK', 'ELEMENT',\n",
       "       'REFERENCE'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_variant_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "#6\n",
    "def row_to_dict(row):\n",
    "    entry_id = row['ENTRY']\n",
    "    name = row['NAME']\n",
    "    gene = row['GENE']\n",
    "    variation = row['VARIATION']\n",
    "    network = row['NETWORK']\n",
    "    element = row['ELEMENT']\n",
    "    reference = row['REFERENCE']\n",
    "    text = f\"name: {name}; gene: {gene}; variation: {variation}; network{network}; element{element}\"\n",
    "\n",
    "    output_dict = {\n",
    "            \"id\": entry_id,\n",
    "            \"input\": text,\n",
    "            \"metadata\": {\n",
    "                \"name\": name,\n",
    "                \"text\": text,\n",
    "                \"network\": network,\n",
    "                \"element\": element,\n",
    "                \"reference\": reference\n",
    "            }\n",
    "        }\n",
    "\n",
    "    return output_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "458\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'id': '10026v1',\n",
       " 'input': 'name: PIGK deficiency; gene: PIGK  phosphatidylinositol glycan anchor biosynthesis class K [KO:K05290]; variation: mutations\\nOmimVar: 605087; networknan; elementnan',\n",
       " 'metadata': {'name': 'PIGK deficiency',\n",
       "  'text': 'name: PIGK deficiency; gene: PIGK  phosphatidylinositol glycan anchor biosynthesis class K [KO:K05290]; variation: mutations\\nOmimVar: 605087; networknan; elementnan',\n",
       "  'network': nan,\n",
       "  'element': nan,\n",
       "  'reference': 'PMID:32220290\\nAUTHORS   Nguyen TTM, Murakami Y, Mobilio S, Niceta M, Zampino G, Philippe C, Moutton S, Zaki MS, James KN, Musaev D, Mu W, Baranano K, Nance JR, Rosenfeld JA, Braverman N, Ciolfi A, Millan F, Person RE, Bruel AL, Thauvin-Robinet C, Ververi A, DeVile C, Male A, Efthymiou S, Maroofian R, Houlden H, Maqbool S, Rahman F, Baratang NV, Rousseau J, St-Denis A, Elrick MJ, Anselm I, Rodan LH, Tartaglia M, Gleeson J, Kinoshita T, Campeau PM\\nTITLE     Bi-allelic Variants in the GPI Transamidase Subunit PIGK Cause a Neurodevelopmental Syndrome with Hypotonia, Cerebellar Atrophy, and Epilepsy.\\nJOURNAL   Am J Hum Genet 106:484-495 (2020)\\nDOI:10.1016/j.ajhg.2020.03.001'}}"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_preprocess = convert_dataframe_to_list(clean_variant_df)\n",
    "print(len(variant_preprocess))\n",
    "variant_preprocess[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(458, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>input</th>\n",
       "      <th>metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10026v1</td>\n",
       "      <td>name: PIGK deficiency; gene: PIGK  phosphatidy...</td>\n",
       "      <td>{'name': 'PIGK deficiency', 'text': 'name: PIG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100v1</td>\n",
       "      <td>name: ADA deficiency; gene: ADA  adenosine dea...</td>\n",
       "      <td>{'name': 'ADA deficiency', 'text': 'name: ADA ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10133v1</td>\n",
       "      <td>name: OPTN mutation; gene: OPTN  optineurin [K...</td>\n",
       "      <td>{'name': 'OPTN mutation', 'text': 'name: OPTN ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10195v1</td>\n",
       "      <td>name: ALG3 deficiency; gene: ALG3  ALG3 alpha-...</td>\n",
       "      <td>{'name': 'ALG3 deficiency', 'text': 'name: ALG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1019v1</td>\n",
       "      <td>name: CDK4 amplification; gene: CDK4  cyclin d...</td>\n",
       "      <td>{'name': 'CDK4 amplification', 'text': 'name: ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id                                              input  \\\n",
       "0  10026v1  name: PIGK deficiency; gene: PIGK  phosphatidy...   \n",
       "1    100v1  name: ADA deficiency; gene: ADA  adenosine dea...   \n",
       "2  10133v1  name: OPTN mutation; gene: OPTN  optineurin [K...   \n",
       "3  10195v1  name: ALG3 deficiency; gene: ALG3  ALG3 alpha-...   \n",
       "4   1019v1  name: CDK4 amplification; gene: CDK4  cyclin d...   \n",
       "\n",
       "                                            metadata  \n",
       "0  {'name': 'PIGK deficiency', 'text': 'name: PIG...  \n",
       "1  {'name': 'ADA deficiency', 'text': 'name: ADA ...  \n",
       "2  {'name': 'OPTN mutation', 'text': 'name: OPTN ...  \n",
       "3  {'name': 'ALG3 deficiency', 'text': 'name: ALG...  \n",
       "4  {'name': 'CDK4 amplification', 'text': 'name: ...  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df = pd.DataFrame(variant_preprocess)\n",
    "preprocessed_datasets.append(variant_df)\n",
    "print(variant_df.shape)\n",
    "variant_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Preprocessing (Creating Embeddings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "index_data = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_embedding(text):\n",
    "    return embed.embed_documents(text)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "indexing (11938, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIError: The server had an error while processing your request. Sorry about that! You can retry your request, or contact us through our help center at help.openai.com if the error persists. (Please include the request ID de0796e13eb9f6a9de50ab52230c7013 in your message.) {\n",
      "  \"error\": {\n",
      "    \"message\": \"The server had an error while processing your request. Sorry about that! You can retry your request, or contact us through our help center at help.openai.com if the error persists. (Please include the request ID de0796e13eb9f6a9de50ab52230c7013 in your message.)\",\n",
      "    \"type\": \"server_error\",\n",
      "    \"param\": null,\n",
      "    \"code\": null\n",
      "  }\n",
      "}\n",
      " 500 {'error': {'message': 'The server had an error while processing your request. Sorry about that! You can retry your request, or contact us through our help center at help.openai.com if the error persists. (Please include the request ID de0796e13eb9f6a9de50ab52230c7013 in your message.)', 'type': 'server_error', 'param': None, 'code': None}} {'Date': 'Sat, 21 Oct 2023 22:49:13 GMT', 'Content-Type': 'application/json', 'Content-Length': '366', 'Connection': 'keep-alive', 'access-control-allow-origin': '*', 'openai-organization': 'telkom-rqoajs', 'openai-processing-ms': '30017', 'openai-version': '2020-10-01', 'strict-transport-security': 'max-age=15724800; includeSubDomains', 'x-ratelimit-limit-requests': '3000', 'x-ratelimit-limit-tokens': '1000000', 'x-ratelimit-remaining-requests': '2999', 'x-ratelimit-remaining-tokens': '999881', 'x-ratelimit-reset-requests': '20ms', 'x-ratelimit-reset-tokens': '7ms', 'x-request-id': 'de0796e13eb9f6a9de50ab52230c7013', 'CF-Cache-Status': 'DYNAMIC', 'Server': 'cloudflare', 'CF-RAY': '819d0cb8c8490392-JNB', 'alt-svc': 'h3=\":443\"; ma=86400'}.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "indexing (2577, 3)\n",
      "indexing (2398, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIConnectionError: Error communicating with OpenAI: HTTPSConnectionPool(host='api.openai.com', port=443): Max retries exceeded with url: /v1/embeddings (Caused by SSLError(SSLEOFError(8, '[SSL: UNEXPECTED_EOF_WHILE_READING] EOF occurred in violation of protocol (_ssl.c:1006)'))).\n",
      "Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIConnectionError: Error communicating with OpenAI: HTTPSConnectionPool(host='api.openai.com', port=443): Max retries exceeded with url: /v1/embeddings (Caused by SSLError(SSLEOFError(8, '[SSL: UNEXPECTED_EOF_WHILE_READING] EOF occurred in violation of protocol (_ssl.c:1006)'))).\n",
      "Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIConnectionError: Error communicating with OpenAI: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')).\n",
      "Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIConnectionError: Error communicating with OpenAI: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')).\n",
      "Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised APIConnectionError: Error communicating with OpenAI: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None)).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "indexing (850, 3)\n",
      "indexing (1241, 3)\n",
      "indexing (458, 3)\n"
     ]
    }
   ],
   "source": [
    "for df in preprocessed_datasets:\n",
    "    print(f\"indexing {df.shape}\")\n",
    "    df['values'] = df['input'].apply(get_embedding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>metadata</th>\n",
       "      <th>values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D00001</td>\n",
       "      <td>{'name': 'Water (JP18/USP);\n",
       "Purified water (JP...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D00002</td>\n",
       "      <td>{'name': 'Nadide (JAN/USAN/INN);\n",
       "Nicotinamide ...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D00003</td>\n",
       "      <td>{'name': 'Oxygen (JP18/USP)', 'text': 'name: O...</td>\n",
       "      <td>[-0.00636011841067576, -0.024926392883966173, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D00004</td>\n",
       "      <td>{'name': 'Carbon dioxide (JP18/USP);\n",
       "Carbon di...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D00005</td>\n",
       "      <td>{'name': 'Flavin adenine dinucleotide (JAN);\n",
       "A...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                           metadata  \\\n",
       "0  D00001  {'name': 'Water (JP18/USP);\n",
       "Purified water (JP...   \n",
       "1  D00002  {'name': 'Nadide (JAN/USAN/INN);\n",
       "Nicotinamide ...   \n",
       "2  D00003  {'name': 'Oxygen (JP18/USP)', 'text': 'name: O...   \n",
       "3  D00004  {'name': 'Carbon dioxide (JP18/USP);\n",
       "Carbon di...   \n",
       "4  D00005  {'name': 'Flavin adenine dinucleotide (JAN);\n",
       "A...   \n",
       "\n",
       "                                              values  \n",
       "0  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "1  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "2  [-0.00636011841067576, -0.024926392883966173, ...  \n",
       "3  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "4  [-0.006311894406848522, -0.02489179188784771, ...  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugs_index = drugs_df.drop(['input'], axis=1)\n",
    "drugs_dataset = pd.DataFrame.to_csv(drugs_index)\n",
    "#with open('index_datasets/drugs.csv', 'w') as f: \n",
    "#    f.write(drugs_dataset)\n",
    "index_data.append(drugs_index)\n",
    "drugs_index.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>metadata</th>\n",
       "      <th>values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H00001</td>\n",
       "      <td>{'name': 'B-cell acute lymphoblastic leukemia;...</td>\n",
       "      <td>[-0.00638153408315556, -0.024893257889223787, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H00002</td>\n",
       "      <td>{'name': 'T-cell acute lymphoblastic leukemia;...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H00003</td>\n",
       "      <td>{'name': 'Acute myeloid leukemia', 'text': 'na...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H00004</td>\n",
       "      <td>{'name': 'Chronic myeloid leukemia', 'text': '...</td>\n",
       "      <td>[-0.006348727372756122, -0.02492023854670328, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H00005</td>\n",
       "      <td>{'name': 'Chronic lymphocytic leukemia', 'text...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                           metadata  \\\n",
       "0  H00001  {'name': 'B-cell acute lymphoblastic leukemia;...   \n",
       "1  H00002  {'name': 'T-cell acute lymphoblastic leukemia;...   \n",
       "2  H00003  {'name': 'Acute myeloid leukemia', 'text': 'na...   \n",
       "3  H00004  {'name': 'Chronic myeloid leukemia', 'text': '...   \n",
       "4  H00005  {'name': 'Chronic lymphocytic leukemia', 'text...   \n",
       "\n",
       "                                              values  \n",
       "0  [-0.00638153408315556, -0.024893257889223787, ...  \n",
       "1  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "2  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "3  [-0.006348727372756122, -0.02492023854670328, ...  \n",
       "4  [-0.006311894406848522, -0.02489179188784771, ...  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_index = disease_df.drop(['input'], axis=1)\n",
    "disease_dataset = pd.DataFrame.to_csv(disease_index)\n",
    "#with open('index_datasets/disease.csv', 'w') as f: \n",
    "#    f.write(disease_dataset)\n",
    "index_data.append(disease_index)\n",
    "disease_index.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>metadata</th>\n",
       "      <th>values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DG00001</td>\n",
       "      <td>{'text': 'name: Chlorhexidine\n",
       " TYPE        Che...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DG00002</td>\n",
       "      <td>{'text': 'name: Oxyquinoline\n",
       " TYPE        Chem...</td>\n",
       "      <td>[-0.006341026706393906, -0.024902941958120584,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DG00003</td>\n",
       "      <td>{'text': 'name: Neomycin\n",
       " ABBR      NEO\n",
       " TYPE ...</td>\n",
       "      <td>[-0.006348727372756122, -0.02492023854670328, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DG00004</td>\n",
       "      <td>{'text': 'name: Miconazole\n",
       " TYPE        Chemic...</td>\n",
       "      <td>[-0.006380894490020517, -0.024877912481013318,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DG00005</td>\n",
       "      <td>{'text': 'name: Tetracycline\n",
       " ABBR      TET\n",
       " T...</td>\n",
       "      <td>[-0.006348727372756122, -0.02492023854670328, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id                                           metadata  \\\n",
       "0  DG00001  {'text': 'name: Chlorhexidine\n",
       " TYPE        Che...   \n",
       "1  DG00002  {'text': 'name: Oxyquinoline\n",
       " TYPE        Chem...   \n",
       "2  DG00003  {'text': 'name: Neomycin\n",
       " ABBR      NEO\n",
       " TYPE ...   \n",
       "3  DG00004  {'text': 'name: Miconazole\n",
       " TYPE        Chemic...   \n",
       "4  DG00005  {'text': 'name: Tetracycline\n",
       " ABBR      TET\n",
       " T...   \n",
       "\n",
       "                                              values  \n",
       "0  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "1  [-0.006341026706393906, -0.024902941958120584,...  \n",
       "2  [-0.006348727372756122, -0.02492023854670328, ...  \n",
       "3  [-0.006380894490020517, -0.024877912481013318,...  \n",
       "4  [-0.006348727372756122, -0.02492023854670328, ...  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgroup_index = dgroup_df.drop(['input'], axis=1)\n",
    "dgroup_dataset = pd.DataFrame.to_csv(dgroup_index)\n",
    "#with open('index_datasets/dgroup.csv', 'w') as f: \n",
    "#   f.write(dgroup_dataset)\n",
    "index_data.append(dgroup_index)\n",
    "dgroup_index.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>metadata</th>\n",
       "      <th>values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E00001</td>\n",
       "      <td>{'name': 'Coptis rhizome (JP17)', 'text': 'nam...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E00002</td>\n",
       "      <td>{'name': 'Evodia fruit (JP17)', 'text': 'name:...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E00003</td>\n",
       "      <td>{'name': 'Saffron (JP17)', 'text': 'name: Saff...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E00004</td>\n",
       "      <td>{'name': 'Agar (JP17/NF)', 'text': 'name: Agar...</td>\n",
       "      <td>[-0.00636011841067576, -0.024926392883966173, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E00005</td>\n",
       "      <td>{'name': 'Tragacanth (JP17/NF)', 'text': 'name...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                           metadata  \\\n",
       "0  E00001  {'name': 'Coptis rhizome (JP17)', 'text': 'nam...   \n",
       "1  E00002  {'name': 'Evodia fruit (JP17)', 'text': 'name:...   \n",
       "2  E00003  {'name': 'Saffron (JP17)', 'text': 'name: Saff...   \n",
       "3  E00004  {'name': 'Agar (JP17/NF)', 'text': 'name: Agar...   \n",
       "4  E00005  {'name': 'Tragacanth (JP17/NF)', 'text': 'name...   \n",
       "\n",
       "                                              values  \n",
       "0  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "1  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "2  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "3  [-0.00636011841067576, -0.024926392883966173, ...  \n",
       "4  [-0.006311894406848522, -0.02489179188784771, ...  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "environ_index = environ_df.drop(['input'], axis=1)\n",
    "environ_dataset = pd.DataFrame.to_csv(environ_index)\n",
    "#with open('index_datasets/environ.csv', 'w') as f: \n",
    "#    f.write(environ_dataset)\n",
    "index_data.append(environ_index)\n",
    "environ_index.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>metadata</th>\n",
       "      <th>values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10026v1</td>\n",
       "      <td>{'name': 'PIGK deficiency', 'text': 'name: PIG...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100v1</td>\n",
       "      <td>{'name': 'ADA deficiency', 'text': 'name: ADA ...</td>\n",
       "      <td>[-0.00636011841067576, -0.024926392883966173, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10133v1</td>\n",
       "      <td>{'name': 'OPTN mutation', 'text': 'name: OPTN ...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10195v1</td>\n",
       "      <td>{'name': 'ALG3 deficiency', 'text': 'name: ALG...</td>\n",
       "      <td>[-0.00636011841067576, -0.024926392883966173, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1019v1</td>\n",
       "      <td>{'name': 'CDK4 amplification', 'text': 'name: ...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        id                                           metadata  \\\n",
       "0  10026v1  {'name': 'PIGK deficiency', 'text': 'name: PIG...   \n",
       "1    100v1  {'name': 'ADA deficiency', 'text': 'name: ADA ...   \n",
       "2  10133v1  {'name': 'OPTN mutation', 'text': 'name: OPTN ...   \n",
       "3  10195v1  {'name': 'ALG3 deficiency', 'text': 'name: ALG...   \n",
       "4   1019v1  {'name': 'CDK4 amplification', 'text': 'name: ...   \n",
       "\n",
       "                                              values  \n",
       "0  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "1  [-0.00636011841067576, -0.024926392883966173, ...  \n",
       "2  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "3  [-0.00636011841067576, -0.024926392883966173, ...  \n",
       "4  [-0.006311894406848522, -0.02489179188784771, ...  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_index = variant_df.drop(['input'], axis=1)\n",
    "variant_dataset = pd.DataFrame.to_csv(variant_index)\n",
    "#with open('index_datasets/variant.csv', 'w') as f: \n",
    "#    f.write(variant_dataset)\n",
    "index_data.append(variant_index)\n",
    "variant_index.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>metadata</th>\n",
       "      <th>values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N00001</td>\n",
       "      <td>{'name': 'EGF-EGFR-RAS-ERK signaling pathway',...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N00002</td>\n",
       "      <td>{'name': 'BCR-ABL fusion kinase to RAS-ERK sig...</td>\n",
       "      <td>[-0.006306162547569985, -0.02492137430814508, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N00003</td>\n",
       "      <td>{'name': 'Mutation-activated KIT to RAS-ERK si...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N00004</td>\n",
       "      <td>{'name': 'Duplication or mutation-activated FL...</td>\n",
       "      <td>[-0.006311894406848522, -0.02489179188784771, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N00005</td>\n",
       "      <td>{'name': 'Mutation-activated MET to RAS-ERK si...</td>\n",
       "      <td>[-0.006306162547569985, -0.02492137430814508, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                           metadata  \\\n",
       "0  N00001  {'name': 'EGF-EGFR-RAS-ERK signaling pathway',...   \n",
       "1  N00002  {'name': 'BCR-ABL fusion kinase to RAS-ERK sig...   \n",
       "2  N00003  {'name': 'Mutation-activated KIT to RAS-ERK si...   \n",
       "3  N00004  {'name': 'Duplication or mutation-activated FL...   \n",
       "4  N00005  {'name': 'Mutation-activated MET to RAS-ERK si...   \n",
       "\n",
       "                                              values  \n",
       "0  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "1  [-0.006306162547569985, -0.02492137430814508, ...  \n",
       "2  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "3  [-0.006311894406848522, -0.02489179188784771, ...  \n",
       "4  [-0.006306162547569985, -0.02492137430814508, ...  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "network_index =  network_df.drop(['input'], axis=1)\n",
    "network_dataset = pd.DataFrame.to_csv(network_index)\n",
    "index_data.append\n",
    "#with open('index_datasets/network.csv', 'w') as f: \n",
    "#    f.write(network_dataset)\n",
    "index_data.append(network_index)\n",
    "network_index.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Storage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "index_name = \"kegg-medicus-database-index\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\katle\\anaconda3\\envs\\nlp_env\\Lib\\site-packages\\pinecone\\index.py:4: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)\n",
      "  from tqdm.autonotebook import tqdm\n"
     ]
    }
   ],
   "source": [
    "import pinecone\n",
    "\n",
    "pinecone_api_key = os.getenv('PINECONE_API_KEY')\n",
    "\n",
    "pinecone.init(\n",
    "    api_key=os.environ['PINECONE_API_KEY'],\n",
    "    environment='gcp-starter'\n",
    ")\n",
    "\n",
    "if index_name not in pinecone.list_indexes():\n",
    "    # we create a new index\n",
    "    pinecone.create_index(\n",
    "        name=index_name,\n",
    "        metric='dotproduct',\n",
    "        dimension= len(result[0])  # 1536 dim of text-embedding-ada-002\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'dimension': 1536,\n",
       " 'index_fullness': 0.0,\n",
       " 'namespaces': {'': {'vector_count': 0}},\n",
       " 'total_vector_count': 0}"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_medicus_index = pinecone.GRPCIndex(index_name)\n",
    "kegg_medicus_index.describe_index_stats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d0456e5178de4db6a23b2ea63824c8d5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "sending upsert requests:   0%|          | 0/11938 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bdec9ea297094699af765a86b23d2bd9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "collecting async responses:   0%|          | 0/120 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "de9b0652241e4d0092b4bab7ef1a7e1d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "sending upsert requests:   0%|          | 0/2577 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9c53f69dd2a44599b1bb02d19bdd521c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "collecting async responses:   0%|          | 0/26 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b1cdbbfc41cb46a7a1c4c094ba41240d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "sending upsert requests:   0%|          | 0/2398 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "621753bb4cb34b1883a83bbfd1e45e96",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "collecting async responses:   0%|          | 0/24 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bf42a4c4b1de403d8f9c1d03f6a5c15a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "sending upsert requests:   0%|          | 0/850 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c7bf6ecdc5b34412b4fed9b0931bd5f6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "collecting async responses:   0%|          | 0/9 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a8d6e9e0a2ac484fae7f91a8c2d7b60c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "sending upsert requests:   0%|          | 0/458 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9d1c8b81b7544ed68a1a95258e126cf3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "collecting async responses:   0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "06c76c2ef0214c4db5d563e6eab3642a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "sending upsert requests:   0%|          | 0/1241 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cfd48328974f414eabca2b05e4b25d9f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "collecting async responses:   0%|          | 0/13 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for df in index_data:\n",
    "    kegg_medicus_index.upsert_from_dataframe(df, batch_size=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'dimension': 1536,\n",
       " 'index_fullness': 0.19462,\n",
       " 'namespaces': {'': {'vector_count': 19462}},\n",
       " 'total_vector_count': 19462}"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_medicus_index.describe_index_stats()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nlp_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
